**Author:** Ronny F. Pacheco
**Date:** Jul 2024  
**Copyright:** © 2024 Ronny Pacheco 
**License:** MIT License

---

MIT License

Copyright (c) 2024 Ronny Pacheco

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


# Needed modules

In [99]:
# Import needed modules
import pandas as pd  # type: ignore
import re  # type: ignore
import os
import pickle
import json


In [14]:
## https://kioku-space.com/en/jupyter-skip-execution/
from IPython.core.magic import register_cell_magic # type: ignore

@register_cell_magic
def skip(line, cell):
    print("Skipping cell")
    if line and cell:
        pass
    return

# Pickle save

In [15]:
%%skip
# =============================================================================
# main function
# =============================================================================
def data_save_load(option):
    """
    This function is used to save or load data for the Jupyter notebook.
    
    Parameters:
    option (str): Either 'save' or 'load' to save or load variables.
    
    Returns:
    dict: Dictionary of loaded variables (only when option is 'load').
    """
    path_folder = "ipynb_db"  # Folder to save variables
    os.makedirs(path_folder, exist_ok=True)  # Create folder if not exist
    path_pickle = os.path.join(path_folder, f"{os.path.basename(globals()['__session__'])}_variables.pkl") # Path to save the variables

    if option == "save":
        with open(path_pickle, "wb") as file:
            variables_dict = {}  # Dictionary to save the variables
            pickle.dump(variables_dict, file)
    elif option == "load":
        with open(path_pickle, "rb") as file:
            variables = pickle.load(file)
        # Now load the variables
        for var_key, var_value in variables.items():
            variables[var_key] = var_value

# =============================================================================
# Call the function
# =============================================================================
data_save_load(option="save")

Skipping cell


# Load_data

In [17]:
# Getting path
path_file = "./data/20240703111001_LINF-Tabla_maestra_v3-2024.xlsx"

# Importing excel data with multiple tabs
df = pd.read_excel(path_file, header=0)

## Checking data

In [18]:
# Checking the table data
print(df.shape)
print(df.dtypes)
df.head()

(9871, 14)
Chrom                 object
Gene_ID               object
Gene_coord            object
Gen_seq               object
Transcript_ID         object
Transcript_coord      object
Transcript_seq        object
CDS/miscRNA_ID        object
Pseudogen             object
CDS/miscRNA_coord     object
CDS/miscRNA_nt_seq    object
CDS_aa_seq            object
Function              object
Notas-JMR             object
dtype: object


Chrom         Gene_ID             Gene_coord  \
0  LinJ.01  LINF_010005000    LinJ.01:1520:5066:-   
1  LinJ.01  LINF_010005100    LinJ.01:5067:7570:-   
2  LinJ.01  LINF_010005200   LinJ.01:8150:11640:-   
3  LinJ.01  LINF_010005300  LinJ.01:11906:12886:-   
4  LinJ.01  LINF_010005400  LinJ.01:13649:17111:-   

                                             Gen_seq    Transcript_ID  \
0  TCCTAGCGGACCCTTGACGCAACCACTGCTCCGCCTTTCTCCTTCT...  LINF_01T0005000   
1  TATTCCCTCTGCAAACGCATCGCTCACACCGACCCACCTACGCACA...  LINF_01T0005100   
2  GGCCGTCTGACGCCCTCCGACCACACCCTCGTCCTCGACAGCACCC...  LINF_01T0005200   
3  CGAGGAGAGGGGCGGACGCACCCGCCCCCCCTCACTCTCTGCACCG...  LINF_01T0005300   
4  CCGCCGTGTCCCAGCTCTGCCTGTGCTCGTCTCTCCCCTCAACCCC...  LINF_01T0005400   

        Transcript_coord                                     Transcript_seq  \
0    LinJ.01:1520:5066:-  TCCTAGCGGACCCTTGACGCAACCACTGCTCCGCCTTTCTCCTTCT...   
1    LinJ.01:5067:7570:-  TATTCCCTCTGCAAACGCATCGCTCACACCGACCCACCTACGCACA...   
2   LinJ.01:8150:11640:-  GGCCGTCTGACGCCCTCCGACCACACCCTCGTCCTCGACAGCACCC...   
3  LinJ.01:11906:12886:-  CGAGGAGAGGGGCGGACGCACCCGCCCCCCCTCACTCTCTGCACCG...   
4  LinJ.01:13649:17111:-  CCGCCGTGTCCCAGCTCTGCCTGTGCTCGTCTCTCCCCTCAACCCC...   

   CDS/miscRNA_ID Pseudogen      CDS/miscRNA_coord  \
0  LINF_010005000        NO    LinJ.01:3710:4711:-   
1  LINF_010005100        NO    LinJ.01:5804:7438:-   
2  LINF_010005200        NO   LinJ.01:9038:11059:-   
3  LINF_010005300        NO  LinJ.01:12041:12601:-   
4  LINF_010005400        NO  LinJ.01:14957:16954:-   

                                  CDS/miscRNA_nt_seq  \
0  ATGCGCTGCGCTCTCGTCTTCGTGCTGGTCGCGGCGGCGCTCGCGT...   
1  ATGTTGAGACGCTGCTCAGCGCTGCTGAATGATCCGCCACTGGTGA...   
2  ATGATGTCGGCCGAGCCGCCGTCGTCGCAGCCGTACATCAGCGACG...   
3  ATGAAGCGAGCGCGTTCTCCGTCGGCCGACAGCACCGTCTCAGCCT...   
4  ATGCCTGCTCCGATTTTCGACAAGGTTCTCGTCGCTAACCGCGGCG...   

                                          CDS_aa_seq  \
0  MRCALVFVLVAAALACFAPAMTHAYSTTYAARSAGSGAFAFDPLEI...   
1  MLRRCSALLNDPPLVSGGPVLGLMRDTGTRTTIRPSIMDAQITQSA...   
2  MMSAEPPSSQPYISDVLRRYQLERFQSSFASSMTIKDLLALQPEDF...   
3  MKRARSPSADSTVSASAPSAVSLSSSAVFDRSLSSSSSLSAIEEKE...   
4  MPAPIFDKVLVANRGEIACRVMATCQRLGIKTVAVYSTADEQAKHV...   

                                      Function Notas-JMR  
0        Protein of unknown function (DUF2946)       NaN  
1  Endonuclease/Exonuclease/phosphatase family       NaN  
2                                   Kinesin-13       NaN  
3             hypothetical protein - conserved       NaN  
4        Propionyl-CoA carboxylase - α-subunit       NaN

For a GTF file we'll need the next columns:
1. SeqID
2. Source
3. Feature
4. Start
5. End
6. Score
7. Strand
8. Phase
9. Attributes

All but the last one need a value. If not value is needed, use the dot ".".

# Cleaning/Preparing data

In [19]:
# Needed columns to make the GTF
needed_columns = ['Chrom', 'Gene_ID', 'Gene_coord', 'Transcript_ID', 'Transcript_coord', 'CDS/miscRNA_ID', 'Pseudogen', 'CDS/miscRNA_coord', 'Function']

In [20]:
# Selecting only the needed columns
df_filtered = df[needed_columns].copy()
print(df_filtered.shape)
print(df_filtered.dtypes)
df_filtered.head()

(9871, 9)
Chrom                object
Gene_ID              object
Gene_coord           object
Transcript_ID        object
Transcript_coord     object
CDS/miscRNA_ID       object
Pseudogen            object
CDS/miscRNA_coord    object
Function             object
dtype: object


Chrom         Gene_ID             Gene_coord    Transcript_ID  \
0  LinJ.01  LINF_010005000    LinJ.01:1520:5066:-  LINF_01T0005000   
1  LinJ.01  LINF_010005100    LinJ.01:5067:7570:-  LINF_01T0005100   
2  LinJ.01  LINF_010005200   LinJ.01:8150:11640:-  LINF_01T0005200   
3  LinJ.01  LINF_010005300  LinJ.01:11906:12886:-  LINF_01T0005300   
4  LinJ.01  LINF_010005400  LinJ.01:13649:17111:-  LINF_01T0005400   

        Transcript_coord  CDS/miscRNA_ID Pseudogen      CDS/miscRNA_coord  \
0    LinJ.01:1520:5066:-  LINF_010005000        NO    LinJ.01:3710:4711:-   
1    LinJ.01:5067:7570:-  LINF_010005100        NO    LinJ.01:5804:7438:-   
2   LinJ.01:8150:11640:-  LINF_010005200        NO   LinJ.01:9038:11059:-   
3  LinJ.01:11906:12886:-  LINF_010005300        NO  LinJ.01:12041:12601:-   
4  LinJ.01:13649:17111:-  LINF_010005400        NO  LinJ.01:14957:16954:-   

                                      Function  
0        Protein of unknown function (DUF2946)  
1  Endonuclease/Exonuclease/phosphatase family  
2                                   Kinesin-13  
3             hypothetical protein - conserved  
4        Propionyl-CoA carboxylase - α-subunit

## Missing, ND, N/A and unknown values

In [21]:
# Check all the columns for missing values, "ND", "N/A", and "unknown"
for column in df_filtered.columns:
    if df_filtered[column].isnull().values.any():
        print(f"{column} has {df_filtered[column].isnull().sum()} <missing> values")
    if "ND" in df_filtered[column].values:
        print(f"{column} has {df_filtered[column].value_counts()['ND']} <ND> values")
    if "N/A" in df_filtered[column].values:
        print(f"{column} has {df_filtered[column].value_counts()['N/A']} <N/A> values")
    if "unknown" in df_filtered[column].values:
        print(f"{column} has {df_filtered[column].value_counts()['unknown']} <unknown> values")

Gene_coord has 1 <ND> values
Transcript_ID has 210 <ND> values
Transcript_coord has 210 <ND> values
CDS/miscRNA_ID has 1103 <ND> values
CDS/miscRNA_coord has 1100 <ND> values
Function has 1116 <unknown> values


<span style="color:white; background-color:red;">THERE IS ONE MISSING VALUE IN GENE_COORD</span>

This element will be skipped in the future code, because without these coordinates we can't make the GTF

In [22]:
# "Gene_coord" is important, so let's check which element has the ND value
print(df_filtered[df_filtered["Gene_coord"] == "ND"])

        Chrom         Gene_ID Gene_coord Transcript_ID Transcript_coord  \
3806  LinJ.22  LINF_220021000         ND            ND               ND   

      CDS/miscRNA_ID Pseudogen        CDS/miscRNA_coord              Function  
3806  LINF_220021000        NO  LinJ.22:664742:665152:-  hypothetical_protein  


In [23]:
# Let's remove "LINF_220021000" it form the dataframe
df_filtered = df_filtered[df_filtered["Gene_ID"] != "LINF_220021000"].copy()

#### 5.1.1 Checking where "Transcript_coord" is "ND"

In [24]:
# Check elements where "Transcript_coord" or "Transcript_ID" is ND
testing_trans_coord = df_filtered[(df_filtered["Transcript_ID"] == "ND") | (df_filtered["Transcript_coord"] == "ND")]
testing_trans_coord

Chrom             Gene_ID                 Gene_coord Transcript_ID  \
152   LinJ.02      LINF_020010500    LinJ.02:235340:235431:+            ND   
160   LinJ.02  LINF_02.SLRNA.0010    LinJ.02:273078:273181:+            ND   
161   LinJ.02  LINF_02.SLRNA.0020    LinJ.02:273519:273622:+            ND   
162   LinJ.02  LINF_02.SLRNA.0030    LinJ.02:273947:274050:+            ND   
163   LinJ.02  LINF_02.SLRNA.0040    LinJ.02:274388:274491:+            ND   
...       ...                 ...                        ...           ...   
9522  LinJ.36      LINF_360051500  LinJ.36:1648079:1648152:-            ND   
9783  LinJ.36      LINF_360074500  LinJ.36:2530973:2531044:+            ND   
9785  LinJ.36      LINF_360074600  LinJ.36:2531835:2531916:-            ND   
9786  LinJ.36      LINF_360074700  LinJ.36:2531978:2532049:-            ND   
9787  LinJ.36      LINF_360074800  LinJ.36:2532123:2532195:+            ND   

     Transcript_coord      CDS/miscRNA_ID Pseudogen  \
152                ND      LINF_020010500        NO   
160                ND  LINF_02.SLRNA.0010        NO   
161                ND  LINF_02.SLRNA.0020        NO   
162                ND  LINF_02.SLRNA.0030        NO   
163                ND  LINF_02.SLRNA.0040        NO   
...               ...                 ...       ...   
9522               ND      LINF_360051500        NO   
9783               ND      LINF_360074500        NO   
9785               ND      LINF_360074600        NO   
9786               ND      LINF_360074700        NO   
9787               ND      LINF_360074800        NO   

              CDS/miscRNA_coord                Function  
152     LinJ.02:235340:235431:+  tRNA-val|Anticodon gac  
160     LinJ.02:273078:273181:+      spliced_leader_RNA  
161     LinJ.02:273519:273622:+      spliced_leader_RNA  
162     LinJ.02:273947:274050:+      spliced_leader_RNA  
163     LinJ.02:274388:274491:+      spliced_leader_RNA  
...                         ...                     ...  
9522  LinJ.36:1648079:1648152:-                tRNA-Gln  
9783  LinJ.36:2530973:2531044:+                tRNA-Cys  
9785  LinJ.36:2531835:2531916:-                tRNA-Leu  
9786  LinJ.36:2531978:2532049:-                tRNA-Met  
9787  LinJ.36:2532123:2532195:+                tRNA-Pro  

[209 rows x 9 columns]

In [25]:
# Check the functions
testing_trans_coord['Function'].value_counts()

Function
spliced_leader_RNA                                 45
snoRNA                                             25
unknown                                            15
tRNA-Leu                                           11
5S rRNA                                            11
spliced_leader-like_RNA                            10
tRNA-Arg                                            9
tRNA-Gly                                            7
tRNA-Thr                                            6
tRNA-Ala                                            6
rRNA                                                6
tRNA-Pro                                            5
tRNA-Val                                            5
tRNA-Ser                                            5
tRNA-Lys                                            5
tRNA-Gln                                            4
tRNA-Asn                                            4
tRNA-Met                                            4
tRNA-Ile           

They are specials RNA mainly, but let's check:
* protein of unknown function - conserved
* tb-292 membrane associated protein-like protein
* hypothetical_protein
* hypothetical protein
* hypothetical protein - conserved
* unknown

In [26]:
conflicted_zone_1 = testing_trans_coord[testing_trans_coord['Function'].isin(["protein of unknown function - conserved", "tb-292 membrane associated protein-like protein", "hypothetical_protein", "hypothetical protein", "hypothetical protein - conserved", "unknown"])]
conflicted_zone_1

Chrom         Gene_ID                 Gene_coord Transcript_ID  \
2227  LinJ.15  LINF_150010400    LinJ.15:186963:187592:+            ND   
2228  LinJ.15  LINF_150010500    LinJ.15:187730:189344:-            ND   
2229  LinJ.15  LINF_150010600    LinJ.15:189599:190309:-            ND   
2230  LinJ.15  LINF_150010700    LinJ.15:190389:191303:-            ND   
2231  LinJ.15  LINF_150010800    LinJ.15:191558:192268:-            ND   
2232  LinJ.15  LINF_150010900    LinJ.15:192348:193262:-            ND   
2233  LinJ.15  LINF_150011000    LinJ.15:193517:194227:-            ND   
2234  LinJ.15  LINF_150011100    LinJ.15:194307:195221:-            ND   
2235  LinJ.15  LINF_150011200    LinJ.15:195476:196186:-            ND   
2236  LinJ.15  LINF_150011300    LinJ.15:196266:197180:-            ND   
2237  LinJ.15  LINF_150011400    LinJ.15:197435:198145:-            ND   
2238  LinJ.15  LINF_150011500    LinJ.15:198225:199139:-            ND   
2239  LinJ.15  LINF_150011600    LinJ.15:199394:200104:-            ND   
2240  LinJ.15  LINF_150011700    LinJ.15:200184:201098:-            ND   
2241  LinJ.15  LINF_150011800    LinJ.15:201353:202063:-            ND   
2242  LinJ.15  LINF_150011900    LinJ.15:202143:203057:-            ND   
2243  LinJ.15  LINF_150012000    LinJ.15:203312:204151:-            ND   
2244  LinJ.15  LINF_150012100    LinJ.15:204741:219545:+            ND   
7793  LinJ.33  LINF_330040700  LinJ.33:1443861:1447517:+            ND   
7794  LinJ.33  LINF_330040800  LinJ.33:1453827:1467362:+            ND   

     Transcript_coord  CDS/miscRNA_ID Pseudogen          CDS/miscRNA_coord  \
2227               ND  LINF_150010400        NO    LinJ.15:186963:187592:+   
2228               ND  LINF_150010500        NO    LinJ.15:187730:189344:-   
2229               ND  LINF_150010600        NO    LinJ.15:189599:190309:-   
2230               ND  LINF_150010700        NO    LinJ.15:190389:191303:-   
2231               ND  LINF_150010800        NO    LinJ.15:191558:192268:-   
2232               ND  LINF_150010900        NO    LinJ.15:192348:193262:-   
2233               ND  LINF_150011000        NO    LinJ.15:193517:194227:-   
2234               ND  LINF_150011100        NO    LinJ.15:194307:195221:-   
2235               ND  LINF_150011200        NO    LinJ.15:195476:196186:-   
2236               ND  LINF_150011300        NO    LinJ.15:196266:197180:-   
2237               ND  LINF_150011400        NO    LinJ.15:197435:198145:-   
2238               ND  LINF_150011500        NO    LinJ.15:198225:199139:-   
2239               ND  LINF_150011600        NO    LinJ.15:199394:200104:-   
2240               ND  LINF_150011700        NO    LinJ.15:200184:201098:-   
2241               ND  LINF_150011800        NO    LinJ.15:201353:202063:-   
2242               ND  LINF_150011900        NO    LinJ.15:202143:203057:-   
2243               ND  LINF_150012000        NO    LinJ.15:203312:204151:-   
2244               ND  LINF_150012100        NO      LinJ.15:204741:219545   
7793               ND  LINF_330040700        NO  LinJ.33:1443861:1447517:+   
7794               ND  LINF_330040800        NO  LinJ.33:1453827:1467362:+   

                                             Function  
2227          protein of unknown function - conserved  
2228                             hypothetical protein  
2229                                          unknown  
2230                                          unknown  
2231                                          unknown  
2232                                          unknown  
2233                                          unknown  
2234                                          unknown  
2235                                          unknown  
2236                                          unknown  
2237                                          unknown  
2238                                          unknown  
2239                                          unknown  
2240                          

Shouldn't they have a transcript?

<span style="color:red;">@JMR told us to remove the CDS/misRNA_coord and note them</span>

Apart from that: Speaking with @Javi and @JMR (16/07/24) this zone is a conflicted zone. Should add to "Notes/Function" sth like - "conflicted_zone_in_study" for example

In [27]:
# Take the "Gene_ID" from the elements in the last code cell
testing_list = testing_trans_coord[testing_trans_coord['Function'].isin(["protein of unknown function - conserved", "tb-292 membrane associated protein-like protein", "hypothetical_protein", "hypothetical protein", "hypothetical protein - conserved", "unknown"])]\
    ["Gene_ID"].tolist()

# Check list
print(testing_list)

['LINF_150010400', 'LINF_150010500', 'LINF_150010600', 'LINF_150010700', 'LINF_150010800', 'LINF_150010900', 'LINF_150011000', 'LINF_150011100', 'LINF_150011200', 'LINF_150011300', 'LINF_150011400', 'LINF_150011500', 'LINF_150011600', 'LINF_150011700', 'LINF_150011800', 'LINF_150011900', 'LINF_150012000', 'LINF_150012100', 'LINF_330040700', 'LINF_330040800']


In [28]:
# Remove in df_filtered, in the "testing_list" the elements "CDS/miscRNA_ID", and "CDS/miscRNA_coord" to "ND" 
df_filtered.loc[df_filtered["Gene_ID"].isin(testing_list), ["CDS/miscRNA_ID", "CDS/miscRNA_coord"]] = "ND"

# To the "Function" column in the `testing_list` we will add to the original text the words: _conflicted_zone_in_study
df_filtered.loc[df_filtered["Gene_ID"].isin(testing_list), "Function"] = df_filtered.loc[df_filtered["Gene_ID"].isin(testing_list), "Function"] + "_conflicted_zone_in_study"


# Check the elements in the "testing_list"
df_filtered[df_filtered["Gene_ID"].isin(testing_list)]

Chrom         Gene_ID                 Gene_coord Transcript_ID  \
2227  LinJ.15  LINF_150010400    LinJ.15:186963:187592:+            ND   
2228  LinJ.15  LINF_150010500    LinJ.15:187730:189344:-            ND   
2229  LinJ.15  LINF_150010600    LinJ.15:189599:190309:-            ND   
2230  LinJ.15  LINF_150010700    LinJ.15:190389:191303:-            ND   
2231  LinJ.15  LINF_150010800    LinJ.15:191558:192268:-            ND   
2232  LinJ.15  LINF_150010900    LinJ.15:192348:193262:-            ND   
2233  LinJ.15  LINF_150011000    LinJ.15:193517:194227:-            ND   
2234  LinJ.15  LINF_150011100    LinJ.15:194307:195221:-            ND   
2235  LinJ.15  LINF_150011200    LinJ.15:195476:196186:-            ND   
2236  LinJ.15  LINF_150011300    LinJ.15:196266:197180:-            ND   
2237  LinJ.15  LINF_150011400    LinJ.15:197435:198145:-            ND   
2238  LinJ.15  LINF_150011500    LinJ.15:198225:199139:-            ND   
2239  LinJ.15  LINF_150011600    LinJ.15:199394:200104:-            ND   
2240  LinJ.15  LINF_150011700    LinJ.15:200184:201098:-            ND   
2241  LinJ.15  LINF_150011800    LinJ.15:201353:202063:-            ND   
2242  LinJ.15  LINF_150011900    LinJ.15:202143:203057:-            ND   
2243  LinJ.15  LINF_150012000    LinJ.15:203312:204151:-            ND   
2244  LinJ.15  LINF_150012100    LinJ.15:204741:219545:+            ND   
7793  LinJ.33  LINF_330040700  LinJ.33:1443861:1447517:+            ND   
7794  LinJ.33  LINF_330040800  LinJ.33:1453827:1467362:+            ND   

     Transcript_coord CDS/miscRNA_ID Pseudogen CDS/miscRNA_coord  \
2227               ND             ND        NO                ND   
2228               ND             ND        NO                ND   
2229               ND             ND        NO                ND   
2230               ND             ND        NO                ND   
2231               ND             ND        NO                ND   
2232               ND             ND        NO                ND   
2233               ND             ND        NO                ND   
2234               ND             ND        NO                ND   
2235               ND             ND        NO                ND   
2236               ND             ND        NO                ND   
2237               ND             ND        NO                ND   
2238               ND             ND        NO                ND   
2239               ND             ND        NO                ND   
2240               ND             ND        NO                ND   
2241               ND             ND        NO                ND   
2242               ND             ND        NO                ND   
2243               ND             ND        NO                ND   
2244               ND             ND        NO                ND   
7793               ND             ND        NO                ND   
7794               ND             ND        NO                ND   

                                               Function  
2227  protein of unknown function - conserved_confli...  
2228      hypothetical protein_conflicted_zone_in_study  
2229                   unknown_conflicted_zone_in_study  
2230                   unknown_conflicted_zone_in_study  
2231                   unknown_conflicted_zone_in_study  
2232                   unknown_conflicted_zone_in_study  
2233                   unknown_conflicted_zone_in_study  
2234                   unknown_conflicted_zone_in_study  
2235                   unknown_conflicted_zone_in_study  
2236                   unknown_conflicted_zone_in_study  
2237                   unknown_conflicted_zone_in_study  
2238                   unknown_conflicted_zone_in_study  
2239                   unknown_conflicted_zone_in_study  
2240                   unknown_conflicted_zone_in_study  
2241                   unknown_conflicted_zone_in_study  
2242                   unknown_conflicted_zone_in_study  
2243                   unknown_conflic

## Checking coordinates

### Wrong coordinates step 1
Normally the coordinates are like <span style="color:orange">LinJ.01:1000:5000:-</span>:
* The fields are "chromosome", "start", "end" and "strand" separated by ":"

In this part we are going to check if every coordinate have the same structure

In [29]:
# Checking which indices doesn't have the coordinate correct structure
df_condition = df_filtered[(df_filtered['CDS/miscRNA_coord'] != "ND") & (df_filtered['Gene_coord'] != "ND")].copy()
condition = df_condition['Gene_coord'].str.split(":").str[3] == df_condition['CDS/miscRNA_coord'].str.split(":").str[3]  # checking if the 3rd element is sense
wrong_coord_1 = df_condition[~condition]['Gene_ID'].tolist()
print(f"Indices with incorrect coordinates structure: {wrong_coord_1}")

Indices with incorrect coordinates structure: ['LINF_260030600', 'LINF_330015900']


In [30]:
# Checking which indices doesn't have the coordinate correct structure
df_condition = df_filtered[(df_filtered['CDS/miscRNA_coord'] != "ND") & (df_filtered['Gene_coord'] != "ND")].copy()
condition = df_condition['Gene_coord'].str.split(":").str[3] == df_condition['CDS/miscRNA_coord'].str.split(":").str[3]  # checking if the 3rd element is sense
wrong_coord_1 = df_condition[~condition]['Gene_ID'].tolist()
print(f"Indices with incorrect coordinates structure: {wrong_coord_1}")

Indices with incorrect coordinates structure: ['LINF_260030600', 'LINF_330015900']


In [31]:
df_condition.loc[
    df_condition['Gene_ID'].isin(wrong_coord_1), 
    ['Gene_ID', 'Gene_coord', 'Transcript_coord', 'CDS/miscRNA_coord', 'Function']
]

Gene_ID               Gene_coord         Transcript_coord  \
4926  LINF_260030600  LinJ.26:974904:977491:+  LinJ.26:974904:977491:+   
7511  LINF_330015900  LinJ.33:351965:353644:+  LinJ.33:351965:353644:+   

            CDS/miscRNA_coord                          Function  
4926  LinJ.26:975665-977149:+              hypothetical protein  
7511  LinJ.33:352089-353393:+  hypothetical protein - conserved

Seems there are some coordinates not separated by ":", and instead by "-" so it doesn't follow the normal structure. We need to correct it for a correct script indexing, for the moment we save this indices

#### Repairing

In [32]:
# Function to apply the regex substitution
def replace_dash_with_colon(s):
    return re.sub(r'(\d+)-(\d+)', r'\1:\2', s)

# Apply the function to the Series
df_filtered.loc[df_filtered['Gene_ID'].isin(wrong_coord_1), 'CDS/miscRNA_coord'] = df_filtered.loc[df_filtered['Gene_ID'].isin(wrong_coord_1), 'CDS/miscRNA_coord'].apply(replace_dash_with_colon)

# Checking the change
df_filtered.loc[df_filtered['Gene_ID'].isin(wrong_coord_1), ['Gene_ID', 'Gene_coord', 'Transcript_coord', 'CDS/miscRNA_coord', 'Function']]

Gene_ID               Gene_coord         Transcript_coord  \
4926  LINF_260030600  LinJ.26:974904:977491:+  LinJ.26:974904:977491:+   
7511  LINF_330015900  LinJ.33:351965:353644:+  LinJ.33:351965:353644:+   

            CDS/miscRNA_coord                          Function  
4926  LinJ.26:975665:977149:+              hypothetical protein  
7511  LinJ.33:352089:353393:+  hypothetical protein - conserved

### Wrong coordinates step 2
If we remember, the correct coordinate structure is <span style="color:orange">LinJ.01:1000:5000:-</span>:
* * The fields are "chromosome", "start", "end" and "strand" separated by ":".

Since there are 4 fields separated by ":", when we split the data by ":" there should be 4 elements. We are going to check if there are indeed 4 elements and where is not

In [33]:
# Check as well in df_condition for the values when .split[":"] is not 4
df_condition = df_filtered[(df_filtered['CDS/miscRNA_coord'] != "ND") & (df_filtered['Gene_coord'] != "ND")].copy()
df_condition['Gene_coord_split'] = df_condition['Gene_coord'].str.split(":")
df_condition['CDS/miscRNA_coord_split'] = df_condition['CDS/miscRNA_coord'].str.split(":")
condition = (df_condition['Gene_coord_split'].str.len() != 4) | (df_condition['CDS/miscRNA_coord_split'].str.len() != 4)
wrong_coord_2 = df_condition[condition]['Gene_ID'].tolist()
print(f"False indices: {wrong_coord_2}")

False indices: []


In [34]:
# Check the elements
df_condition.loc[
    df_condition['Gene_ID'].isin(wrong_coord_2), 
    ['Gene_ID', 'Gene_coord', 'Transcript_coord', 'CDS/miscRNA_coord', 'Function']
]

Empty DataFrame
Columns: [Gene_ID, Gene_coord, Transcript_coord, CDS/miscRNA_coord, Function]
Index: []

There are no wrong elements in "CDS/miscRNA_coord" and "Gene_coord"

### Wrong coordinates step 3

What if there were an structure like <span style="color:orange">LinJ.01:1000-5000:10000-20000:-</span>?

There would be 4 fields when we separate it by ":":
* LinJ.01
* 1000-5000
* 10000-20000
* "-"

But in the last script, there will be a problem since the start coordinates can't be "1000-5000", so let's check if this exists in the table

In [35]:
# search the regex condition where "number"-"number" in the Gene_coord and CDS/miscRNA_coord columns
# import re
df_condition = df_filtered[(df_filtered['CDS/miscRNA_coord'] != "ND") & (df_filtered['Gene_coord'] != "ND")].copy()
# condition = df_condition['Gene_coord'].str.contains(r"\d+-\d+") | df_condition['CDS/miscRNA_coord'].str.contains(r"\d+-\d+")
condition = df_condition['Gene_coord'].str.contains(r"\d+-\d+") | df_condition['Transcript_coord'].str.contains(r"\d+-\d+") | df_condition['CDS/miscRNA_coord'].str.contains(r"\d+-\d+")

wrong_coord_3 = df_condition[condition]['Gene_ID'].tolist()
print(f"False indices: {wrong_coord_3}")

False indices: ['LINF_070008800', 'LINF_100016800', 'LINF_240023900', 'LINF_290033800']


In [36]:
df_condition.loc[
    df_condition['Gene_ID'].isin(wrong_coord_3), 
    ['Gene_ID', 'Gene_coord', 'Transcript_coord', 'CDS/miscRNA_coord', 'Function']
]

Gene_ID                 Gene_coord  \
842   LINF_070008800    LinJ.07:153281:160411:-   
1431  LINF_100016800    LinJ.10:468436:469642:-   
4288  LINF_240023900    LinJ.24:689866:693427:-   
6002  LINF_290033800  LinJ.29:1181710:1184611:+   

                               Transcript_coord  \
842       LinJ.07:153281-154334:158854-160411:-   
1431      LinJ.10:468436-468904:469155-469642:-   
4288      LinJ.24:689866-690901:692650-693427:-   
6002  LinJ.29:1181710-1182263:1183113-1184611:+   

                              CDS/miscRNA_coord  \
842       LinJ.07:153994-154334:158854-160240:-   
1431                    LinJ.10:469169:469561:-   
4288      LinJ.24:690626-690901:692650-693288:-   
6002  LinJ.29:1182001-1182263:1183113-1184352:+   

                                               Function  
842                  ATP-dependent RNA helicase (DBP2B)  
1431                                         histone H3  
4288                                 nPABP-like protein  
6002  polynucleotide adenylyltransferase|poly(A) pol...

In cases like LINF_070008800, LINF_240023900, LINF_290033800 there seems to be 2 transcripts and 2 CDS.
In the case of LINF_100016800 there 2 transcript but one CDS

We'll have to improve the original table, so it can be parsed later, First let's repair the coordinates:

#### Repairing

In [37]:
# Repair "Transcript_coord"
import re

for elem in wrong_coord_3:
    df_filtered.loc[df_filtered['Gene_ID'] == elem, 'Transcript_coord'] = df_filtered.loc[df_filtered['Gene_ID'] == elem, 'Transcript_coord'].apply(
        lambda x: re.sub(r'(\d+)-(\d+):(\d+)-(\d+)', r'\1:\2-\3:\4', x) if isinstance(x, str) else x
    )

# Check the elements
df_filtered[df_filtered['Gene_ID'].isin(wrong_coord_3)]

Chrom         Gene_ID                 Gene_coord    Transcript_ID  \
842   LinJ.07  LINF_070008800    LinJ.07:153281:160411:-  LINF_07T0008800   
1431  LinJ.10  LINF_100016800    LinJ.10:468436:469642:-  LINF_10T0016800   
4288  LinJ.24  LINF_240023900    LinJ.24:689866:693427:-  LINF_24T0023900   
6002  LinJ.29  LINF_290033800  LinJ.29:1181710:1184611:+  LINF_29T0033800   

                               Transcript_coord  CDS/miscRNA_ID Pseudogen  \
842       LinJ.07:153281:154334-158854:160411:-  LINF_070008800        NO   
1431      LinJ.10:468436:468904-469155:469642:-  LINF_100016800        NO   
4288      LinJ.24:689866:690901-692650:693427:-  LINF_240023900        NO   
6002  LinJ.29:1181710:1182263-1183113:1184611:+  LINF_290033800        NO   

                              CDS/miscRNA_coord  \
842       LinJ.07:153994-154334:158854-160240:-   
1431                    LinJ.10:469169:469561:-   
4288      LinJ.24:690626-690901:692650-693288:-   
6002  LinJ.29:1182001-1182263:1183113-1184352:+   

                                               Function  
842                  ATP-dependent RNA helicase (DBP2B)  
1431                                         histone H3  
4288                                 nPABP-like protein  
6002  polynucleotide adenylyltransferase|poly(A) pol...

In [38]:
# Repairing CDS/miscRNA_coord
cds_coord_repaired = ["LINF_070008800", "LINF_240023900", "LINF_290033800"]
for elem in cds_coord_repaired:
    df_filtered.loc[df_filtered['Gene_ID'] == elem, 'CDS/miscRNA_coord'] = df_filtered.loc[df_filtered['Gene_ID'] == elem, 'CDS/miscRNA_coord'].apply(
        lambda x: re.sub(r'(\d+)-(\d+):(\d+)-(\d+)', r'\1:\2-\3:\4', x) if isinstance(x, str) else x
    )

# Check the elements
df_filtered[df_filtered['Gene_ID'].isin(cds_coord_repaired)]

Chrom         Gene_ID                 Gene_coord    Transcript_ID  \
842   LinJ.07  LINF_070008800    LinJ.07:153281:160411:-  LINF_07T0008800   
4288  LinJ.24  LINF_240023900    LinJ.24:689866:693427:-  LINF_24T0023900   
6002  LinJ.29  LINF_290033800  LinJ.29:1181710:1184611:+  LINF_29T0033800   

                               Transcript_coord  CDS/miscRNA_ID Pseudogen  \
842       LinJ.07:153281:154334-158854:160411:-  LINF_070008800        NO   
4288      LinJ.24:689866:690901-692650:693427:-  LINF_240023900        NO   
6002  LinJ.29:1181710:1182263-1183113:1184611:+  LINF_290033800        NO   

                              CDS/miscRNA_coord  \
842       LinJ.07:153994:154334-158854:160240:-   
4288      LinJ.24:690626:690901-692650:693288:-   
6002  LinJ.29:1182001:1182263-1183113:1184352:+   

                                               Function  
842                  ATP-dependent RNA helicase (DBP2B)  
4288                                 nPABP-like protein  
6002  polynucleotide adenylyltransferase|poly(A) pol...

Now let's divide the row, 2 rows for each transcript.

In [39]:
for gene in cds_coord_repaired:
    # Get column values:
    chrom = df_filtered[df_filtered['Gene_ID'] == gene]['Chrom'].values[0]
    sense = df_filtered[df_filtered['Gene_ID'] == gene]['Gene_coord'].str.split(":").str[3].values[0]
    gene_id = df_filtered[df_filtered['Gene_ID'] == gene]['Gene_ID'].values[0]
    gene_coord = df_filtered[df_filtered['Gene_ID'] == gene]['Gene_coord'].values[0]
    transcript_id = df_filtered[df_filtered['Gene_ID'] == gene]['Transcript_ID'].values[0]
    transcript_coord = df_filtered[df_filtered['Gene_ID'] == gene]['Transcript_coord'].values[0]
    cds_id = df_filtered[df_filtered['Gene_ID'] == gene]['CDS/miscRNA_ID'].values[0]
    pseudogen = df_filtered[df_filtered['Gene_ID'] == gene]['Pseudogen'].values[0]
    cds_coord = df_filtered[df_filtered['Gene_ID'] == gene]['CDS/miscRNA_coord'].values[0]
    function = df_filtered[df_filtered['Gene_ID'] == gene]['Function'].values[0]


    # Get the coordinates
    first_trans_coord = transcript_coord.split("-")[0].split(":")[1] + ":" + transcript_coord.split("-")[0].split(":")[2]
    second_trans_coord = transcript_coord.split("-")[1].split(":")[0] + ":" + transcript_coord.split("-")[1].split(":")[1]

    first_cds_coord = cds_coord.split("-")[0].split(":")[1] + ":" + cds_coord.split("-")[0].split(":")[2]
    second_cds_coord = cds_coord.split("-")[1].split(":")[0] + ":" + cds_coord.split("-")[1].split(":")[1]

    # get the rows:
    first_row = [
        chrom,
        gene_id,
        gene_coord,
        transcript_id + "_1",
        chrom + ":" + first_trans_coord + ":" + sense,
        cds_id + "_1",
        pseudogen,
        chrom + ":" + first_cds_coord + ":" + sense,
        function
    ]

    second_row = [
        chrom,
        gene_id,
        gene_coord,
        transcript_id + "_2",
        chrom + ":" + second_trans_coord + ":" + sense,
        cds_id + "_2",
        pseudogen,
        chrom + ":" + second_cds_coord + ":" + sense,
        function
    ]

    # Transforming the rows to dataframes
    first_row = pd.DataFrame([first_row], columns=df_filtered.columns)
    second_row = pd.DataFrame([second_row], columns=df_filtered.columns)

    # Get the index of the row:
    index_split = df_filtered[df_filtered['Gene_ID'] == gene].index[0]

    # Split the data frame at the index_split
    df_part1 = df_filtered.iloc[:index_split, :]
    df_part2 = df_filtered.iloc[index_split+1:, :]

    # Removing the last row of the first part, because it will be divided
    df_part1 = df_part1.iloc[:-1, :]

    # Inserting the new two rows
    df_filtered = pd.concat([df_part1, first_row, second_row, df_part2], ignore_index=True)

In [40]:
# Let's check the results
df_filtered[df_filtered['Gene_ID'].isin(cds_coord_repaired)]

Chrom         Gene_ID                 Gene_coord      Transcript_ID  \
841   LinJ.07  LINF_070008800    LinJ.07:153281:160411:-  LINF_07T0008800_1   
842   LinJ.07  LINF_070008800    LinJ.07:153281:160411:-  LINF_07T0008800_2   
4286  LinJ.24  LINF_240023900    LinJ.24:689866:693427:-  LINF_24T0023900_1   
4287  LinJ.24  LINF_240023900    LinJ.24:689866:693427:-  LINF_24T0023900_2   
6000  LinJ.29  LINF_290033800  LinJ.29:1181710:1184611:+  LINF_29T0033800_1   
6001  LinJ.29  LINF_290033800  LinJ.29:1181710:1184611:+  LINF_29T0033800_2   

               Transcript_coord    CDS/miscRNA_ID Pseudogen  \
841     LinJ.07:153281:154334:-  LINF_070008800_1        NO   
842     LinJ.07:158854:160411:-  LINF_070008800_2        NO   
4286    LinJ.24:689866:690901:-  LINF_240023900_1        NO   
4287    LinJ.24:692650:693427:-  LINF_240023900_2        NO   
6000  LinJ.29:1181710:1182263:+  LINF_290033800_1        NO   
6001  LinJ.29:1183113:1184611:+  LINF_290033800_2        NO   

              CDS/miscRNA_coord  \
841     LinJ.07:153994:154334:-   
842     LinJ.07:158854:160240:-   
4286    LinJ.24:690626:690901:-   
4287    LinJ.24:692650:693288:-   
6000  LinJ.29:1182001:1182263:+   
6001  LinJ.29:1183113:1184352:+   

                                               Function  
841                  ATP-dependent RNA helicase (DBP2B)  
842                  ATP-dependent RNA helicase (DBP2B)  
4286                                 nPABP-like protein  
4287                                 nPABP-like protein  
6000  polynucleotide adenylyltransferase|poly(A) pol...  
6001  polynucleotide adenylyltransferase|poly(A) pol...

In [41]:
df_filtered[df_filtered['Gene_ID'] == "LINF_100016800"]

Chrom         Gene_ID               Gene_coord    Transcript_ID  \
1431  LinJ.10  LINF_100016800  LinJ.10:468436:469642:-  LINF_10T0016800   

                           Transcript_coord  CDS/miscRNA_ID Pseudogen  \
1431  LinJ.10:468436:468904-469155:469642:-  LINF_100016800        NO   

            CDS/miscRNA_coord    Function  
1431  LinJ.10:469169:469561:-  histone H3

Seems in **LINF_100016800** the CDS (469169:469561) is inside the second transcript (469155:469642) only

In [42]:
# Let's reuse the code 
for gene in ["LINF_100016800"]:
    # Get column values:
    chrom = df_filtered[df_filtered['Gene_ID'] == gene]['Chrom'].values[0]
    sense = df_filtered[df_filtered['Gene_ID'] == gene]['Gene_coord'].str.split(":").str[3].values[0]
    gene_id = df_filtered[df_filtered['Gene_ID'] == gene]['Gene_ID'].values[0]
    gene_coord = df_filtered[df_filtered['Gene_ID'] == gene]['Gene_coord'].values[0]
    transcript_id = df_filtered[df_filtered['Gene_ID'] == gene]['Transcript_ID'].values[0]
    transcript_coord = df_filtered[df_filtered['Gene_ID'] == gene]['Transcript_coord'].values[0]
    cds_id = df_filtered[df_filtered['Gene_ID'] == gene]['CDS/miscRNA_ID'].values[0]
    pseudogen = df_filtered[df_filtered['Gene_ID'] == gene]['Pseudogen'].values[0]
    cds_coord = df_filtered[df_filtered['Gene_ID'] == gene]['CDS/miscRNA_coord'].values[0]
    function = df_filtered[df_filtered['Gene_ID'] == gene]['Function'].values[0]


    # Get the coordinates
    first_trans_coord = transcript_coord.split("-")[0].split(":")[1] + ":" + transcript_coord.split("-")[0].split(":")[2]
    second_trans_coord = transcript_coord.split("-")[1].split(":")[0] + ":" + transcript_coord.split("-")[1].split(":")[1]

    second_cds_coord = df_filtered[df_filtered['Gene_ID'] == gene]['CDS/miscRNA_coord'].values[0]  # This is the only one that changes in the code and it keeps the same value

    # get the rows:
    first_row = [
        chrom,
        gene_id,
        gene_coord,
        transcript_id + "_1",
        chrom + ":" + first_trans_coord + ":" + sense,
        "ND",  # No CDS in this transcript
        pseudogen,
        "ND",  # No CDS in this transcript,
        function
    ]

    second_row = [
        chrom,
        gene_id,
        gene_coord,
        transcript_id + "_2",
        chrom + ":" + second_trans_coord + ":" + sense,
        cds_id, # only one CDS in this gene
        pseudogen,
        second_cds_coord,  # only one CDS in this gene
        function
    ]

    # Transforming the rows to dataframes
    first_row = pd.DataFrame([first_row], columns=df_filtered.columns)
    second_row = pd.DataFrame([second_row], columns=df_filtered.columns)

    # Get the index of the row:
    index_split = df_filtered[df_filtered['Gene_ID'] == gene].index[0]

    # Split the data frame at the index_split
    df_part1 = df_filtered.iloc[:index_split, :]
    df_part2 = df_filtered.iloc[index_split+1:, :]

    # Removing the last row of the first part, because it will be divided
    df_part1 = df_part1.iloc[:-1, :]

    # Inserting the new two rows
    df_filtered = pd.concat([df_part1, first_row, second_row, df_part2], ignore_index=True)

In [43]:
df_filtered[df_filtered['Gene_ID'] == "LINF_100016800"]

Chrom         Gene_ID               Gene_coord      Transcript_ID  \
1430  LinJ.10  LINF_100016800  LinJ.10:468436:469642:-  LINF_10T0016800_1   
1431  LinJ.10  LINF_100016800  LinJ.10:468436:469642:-  LINF_10T0016800_2   

             Transcript_coord  CDS/miscRNA_ID Pseudogen  \
1430  LinJ.10:468436:468904:-              ND        NO   
1431  LinJ.10:469155:469642:-  LINF_100016800        NO   

            CDS/miscRNA_coord    Function  
1430                       ND  histone H3  
1431  LinJ.10:469169:469561:-  histone H3

---

# Other cleaning/preparing data

## Check other possible problems

### Pseudogen categories

In [44]:
# Need to bear in mind the field "pseudogen"
df_filtered['Pseudogen'].value_counts()

Pseudogen
NO     9821
YES      40
SI        9
Name: count, dtype: int64

In the case of "pseudogene" we need to have in mind that there are no only "YES" values as expected, but "SI" values as well.

### CDS/miscRNA with ND but with coordinates

In [45]:
# In this case, there is no ID for the CDS/miscRNA, but there are coordinates. Let's check them
df_filtered[(df_filtered['CDS/miscRNA_ID'] == "ND") & (df_filtered['CDS/miscRNA_coord'] != "ND")]

Chrom         Gene_ID                 Gene_coord    Transcript_ID  \
1462  LinJ.10  LINF_100019760    LinJ.10:549490:550353:-  LINF_10T0019760   
1789  LinJ.12  LINF_120017725    LinJ.12:590295:591611:+  LINF_12T0017725   
9702  LinJ.36  LINF_360067550  LinJ.36:2250628:2251942:-  LINF_36T0067550   

               Transcript_coord CDS/miscRNA_ID Pseudogen  \
1462    LinJ.10:549490:550353:-             ND        NO   
1789    LinJ.12:590295:591611:+             ND        NO   
9702  LinJ.36:2250628:2251942:-             ND        NO   

              CDS/miscRNA_coord                                 Function  
1462    LinJ.10:549923:550327:-  Protein of unknown function - conserved  
1789    LinJ.12:590379:591341:+                        methyltransferase  
9702  LinJ.36:2251654:2251797:-              protein of unknown function

Speaking with @Javi this seems to be an annotation error, let's add the "Gene_ID" column ID to the "CDS/miscRNA_ID"

In [46]:
# For the cases in the cell before, use the "Gene_ID" value in "CDS/miscRNA_ID"
condition = (df_filtered['CDS/miscRNA_ID'] == "ND") & (df_filtered['CDS/miscRNA_coord'] != "ND")
df_filtered.loc[condition, 'CDS/miscRNA_ID'] = df_filtered.loc[condition, 'Gene_ID']
df_filtered.loc[condition]

Chrom         Gene_ID                 Gene_coord    Transcript_ID  \
1462  LinJ.10  LINF_100019760    LinJ.10:549490:550353:-  LINF_10T0019760   
1789  LinJ.12  LINF_120017725    LinJ.12:590295:591611:+  LINF_12T0017725   
9702  LinJ.36  LINF_360067550  LinJ.36:2250628:2251942:-  LINF_36T0067550   

               Transcript_coord  CDS/miscRNA_ID Pseudogen  \
1462    LinJ.10:549490:550353:-  LINF_100019760        NO   
1789    LinJ.12:590295:591611:+  LINF_120017725        NO   
9702  LinJ.36:2250628:2251942:-  LINF_360067550        NO   

              CDS/miscRNA_coord                                 Function  
1462    LinJ.10:549923:550327:-  Protein of unknown function - conserved  
1789    LinJ.12:590379:591341:+                        methyltransferase  
9702  LinJ.36:2251654:2251797:-              protein of unknown function

Now it's repaired

### "CDS/miscRNA_coord" are not inside the coordinates in "Gene_coord"

In [47]:
df_condition[df_condition['CDS/miscRNA_coord'].str.split(":").str[1].values == "+"]

Empty DataFrame
Columns: [Chrom, Gene_ID, Gene_coord, Transcript_ID, Transcript_coord, CDS/miscRNA_ID, Pseudogen, CDS/miscRNA_coord, Function]
Index: []

In [48]:
# Let's check where the coordinates  in "CDS/miscRNA_coord" are not inside the coordinates in "Gene_coord"
## Prepare Data Frame
condition = (df_filtered['CDS/miscRNA_coord'] != "ND") & (df_filtered['Gene_coord'] != "ND")  # Condition to filter, for when the coordinates are not ND
df_condition = df_filtered[condition].copy()
df_condition = df_condition.drop([842, 4288, 6002])  # Drop the rows with the LinJ.XX:100:1000-2000:3000 coordinates types

# Prepare columns
df_condition['Gene_coord_start'] = df_condition['Gene_coord'].str.split(":").str[1].astype(int)
df_condition['Gene_coord_end'] = df_condition['Gene_coord'].str.split(":").str[2].astype(int)
df_condition['CDS/miscRNA_coord_start'] = df_condition['CDS/miscRNA_coord'].str.split(":").str[1].astype(int)
df_condition['CDS/miscRNA_coord_end'] = df_condition['CDS/miscRNA_coord'].str.split(":").str[2].astype(int)
# print(df_condition[['Gene_ID', 'Gene_coord', 'CDS/miscRNA_coord', 'Gene_coord_start', 'Gene_coord_end', 'CDS/miscRNA_coord_start', 'CDS/miscRNA_coord_end']])  # Check the values

# Now let's check that all the coordinates in "CDS/miscRNA_coord" are inside the coordinates in "Gene_coord"
condition = (
    (df_condition['CDS/miscRNA_coord_start'] >= df_condition['Gene_coord_start']) & (df_condition['CDS/miscRNA_coord_start'] <= df_condition['Gene_coord_end']) &
    (df_condition['CDS/miscRNA_coord_end'] >= df_condition['Gene_coord_start']) & (df_condition['CDS/miscRNA_coord_end'] <= df_condition['Gene_coord_end'])
    )

# check the lines that don't fulfill the condition
df_condition[~condition]


Chrom         Gene_ID                 Gene_coord    Transcript_ID  \
1551  LinJ.11  LINF_110011400    LinJ.11:202887:204220:+  LINF_11T0011400   
2067  LinJ.14  LINF_140012250    LinJ.14:248820:251321:-  LINF_14T0012250   
2174  LinJ.15  LINF_150005000        LinJ.15:1108:1687:-  LINF_15T0005000   
3949  LinJ.23  LINF_230015300    LinJ.23:366240:369541:-  LINF_23T0015300   
5906  LinJ.29  LINF_290024900    LinJ.29:850028:850959:-  LINF_29T0024900   
8710  LinJ.35  LINF_350034200  LinJ.35:1106085:1108479:-  LINF_35T0034200   
8711  LinJ.35  LINF_350034300  LinJ.35:1108768:1111622:-  LINF_35T0034300   

               Transcript_coord  CDS/miscRNA_ID Pseudogen  \
1551    LinJ.11:202887:204220:+  LINF_110011400       YES   
2067    LinJ.14:248820:251321:-  LINF_140012250       YES   
2174        LinJ.15:1108:1687:-  LINF_150005000        NO   
3949    LinJ.23:366240:369541:-  LINF_230015300        NO   
5906    LinJ.29:850028:850959:-  LINF_290024900        NO   
8710  LinJ.35:1106085:1108479:-  LINF_350034200        NO   
8711  LinJ.35:1108768:1111622:-  LINF_350034300        NO   

              CDS/miscRNA_coord                                     Function  \
1551    LinJ.11:202513:204017:+          argonaute like protein (pseudogene)   
2067    LinJ.14:250151:251383:-                          fatty acid elongase   
2174        LinJ.15:1505:1715:-                                   histone H4   
3949    LinJ.23:367028:369919:-             hypothetical protein - conserved   
5906    LinJ.29:849730:850788:-               paraflagellar rod protein-like   
8710  LinJ.35:1109345:1111528:-  mt-LAF5|mitoribosomal LSU assembly factor 5   
8711  LinJ.35:1109345:1111843:-             Hypothetical protein - conserved   

      Gene_coord_start  Gene_coord_end  CDS/miscRNA_coord_start  \
1551            202887          204220                   202513   
2067            248820          251321                   250151   
2174              1108            1687                     1505   
3949            366240          369541                   367028   
5906            850028          850959                   849730   
8710           1106085         1108479                  1109345   
8711           1108768         1111622                  1109345   

      CDS/miscRNA_coord_end  
1551                 204017  
2067                 251383  
2174                   1715  
3949                 369919  
5906                 850788  
8710                1111528  
8711                1111843

At the end, we'll have to ignore:
* LINF_360065300 ==> spoke with @Javi and @JMR, hypothetical bi-cistronic gene.
* LINF_110011400, LINF_140012250, LINF_150005000, LINF_230015300, LINF_290024900, LINF_350034200, LINF_350034300  ==> CDS/miscRNA_coord	are outside Gene_coord.
* Elements where `['Gene_coord'] == "ND"` (Already dropped from the table)

### Transcript_coord diff than gene_coord

Normally elements in **Transcript_coord** are the same as in **Gene_coord**, except when a gene has more than 1 transcript: let's check it

In [49]:
test_df = df_filtered.copy()
test_df = test_df[test_df['Transcript_coord'] != "ND"].copy()
test_df[test_df['Gene_coord'] != test_df['Transcript_coord']]

Chrom         Gene_ID                 Gene_coord      Transcript_ID  \
841   LinJ.07  LINF_070008800    LinJ.07:153281:160411:-  LINF_07T0008800_1   
842   LinJ.07  LINF_070008800    LinJ.07:153281:160411:-  LINF_07T0008800_2   
1430  LinJ.10  LINF_100016800    LinJ.10:468436:469642:-  LINF_10T0016800_1   
1431  LinJ.10  LINF_100016800    LinJ.10:468436:469642:-  LINF_10T0016800_2   
4286  LinJ.24  LINF_240023900    LinJ.24:689866:693427:-  LINF_24T0023900_1   
4287  LinJ.24  LINF_240023900    LinJ.24:689866:693427:-  LINF_24T0023900_2   
6000  LinJ.29  LINF_290033800  LinJ.29:1181710:1184611:+  LINF_29T0033800_1   
6001  LinJ.29  LINF_290033800  LinJ.29:1181710:1184611:+  LINF_29T0033800_2   
9840  LinJ.36  LINF_360079700  LinJ.36:2681995:2683375:-    LINF_36T0079700   

               Transcript_coord    CDS/miscRNA_ID Pseudogen  \
841     LinJ.07:153281:154334:-  LINF_070008800_1        NO   
842     LinJ.07:158854:160411:-  LINF_070008800_2        NO   
1430    LinJ.10:468436:468904:-                ND        NO   
1431    LinJ.10:469155:469642:-    LINF_100016800        NO   
4286    LinJ.24:689866:690901:-  LINF_240023900_1        NO   
4287    LinJ.24:692650:693427:-  LINF_240023900_2        NO   
6000  LinJ.29:1181710:1182263:+  LINF_290033800_1        NO   
6001  LinJ.29:1183113:1184611:+  LINF_290033800_2        NO   
9840  LinJ.36:2681395:2683433:-    LINF_360079700        SI   

              CDS/miscRNA_coord  \
841     LinJ.07:153994:154334:-   
842     LinJ.07:158854:160240:-   
1430                         ND   
1431    LinJ.10:469169:469561:-   
4286    LinJ.24:690626:690901:-   
4287    LinJ.24:692650:693288:-   
6000  LinJ.29:1182001:1182263:+   
6001  LinJ.29:1183113:1184352:+   
9840  LinJ.36:2682378:2682647:-   

                                               Function  
841                  ATP-dependent RNA helicase (DBP2B)  
842                  ATP-dependent RNA helicase (DBP2B)  
1430                                         histone H3  
1431                                         histone H3  
4286                                 nPABP-like protein  
4287                                 nPABP-like protein  
6000  polynucleotide adenylyltransferase|poly(A) pol...  
6001  polynucleotide adenylyltransferase|poly(A) pol...  
9840                                               SET8

We already checked all  of them except one:
* LINF_360079700 ==> Strange case. Because the transcript goes beyond the gene coord. Well check it out later

### Duplicate "Gene_ID"

In [50]:
# Check elements where the "Gene_ID" value_counts() is greater than 1
df_filtered['Gene_ID'].value_counts()[df_filtered['Gene_ID'].value_counts() > 1]

Gene_ID
LINF_320037300-400     2
LINF_320029000-100     2
LINF_100016800         2
LINF_290033800         2
LINF_120014500-600     2
LINF_260029800-900     2
LINF_310033100-3000    2
LINF_220018100-200     2
LINF_070008800         2
LINF_320030400-500     2
LINF_100011400-500     2
LINF_250017700-600     2
LINF_240023900         2
LINF_330021700-800     2
LINF_250015300-400     2
LINF_060017800-900     2
LINF_270033600-700     2
LINF_120014100-200     2
LINF_310035100-5000    2
Name: count, dtype: int64

All of  them are by pairs

In [51]:
# df_filtered['Gene_ID'].value_counts()[df_filtered['Gene_ID'].value_counts() > 1]
df_filtered[df_filtered['Gene_ID'].value_counts()[df_filtered['Gene_ID']].values > 1]

Chrom              Gene_ID                 Gene_coord  \
777   LinJ.06   LINF_060017800-900    LinJ.06:487685:490191:+   
778   LinJ.06   LINF_060017800-900    LinJ.06:487685:490191:+   
841   LinJ.07       LINF_070008800    LinJ.07:153281:160411:-   
842   LinJ.07       LINF_070008800    LinJ.07:153281:160411:-   
1376  LinJ.10   LINF_100011400-500    LinJ.10:263964:266643:+   
1377  LinJ.10   LINF_100011400-500    LinJ.10:263964:266643:+   
1430  LinJ.10       LINF_100016800    LinJ.10:468436:469642:-   
1431  LinJ.10       LINF_100016800    LinJ.10:468436:469642:-   
1748  LinJ.12   LINF_120014100-200    LinJ.12:440484:443559:+   
1749  LinJ.12   LINF_120014100-200    LinJ.12:440484:443559:+   
1752  LinJ.12   LINF_120014500-600    LinJ.12:450370:452072:+   
1753  LinJ.12   LINF_120014500-600    LinJ.12:450370:452072:+   
3774  LinJ.22   LINF_220018100-200    LinJ.22:557968:563169:+   
3775  LinJ.22   LINF_220018100-200    LinJ.22:557968:563169:+   
4286  LinJ.24       LINF_240023900    LinJ.24:689866:693427:-   
4287  LinJ.24       LINF_240023900    LinJ.24:689866:693427:-   
4469  LinJ.25   LINF_250015300-400    LinJ.25:347501:349146:+   
4470  LinJ.25   LINF_250015300-400    LinJ.25:347501:349146:+   
4495  LinJ.25   LINF_250017700-600    LinJ.25:451512:453695:-   
4496  LinJ.25   LINF_250017700-600    LinJ.25:451512:453695:-   
4917  LinJ.26   LINF_260029800-900    LinJ.26:953645:955999:+   
4918  LinJ.26   LINF_260029800-900    LinJ.26:953645:955999:+   
5286  LinJ.27   LINF_270033600-700  LinJ.27:1153377:1155197:+   
5287  LinJ.27   LINF_270033600-700  LinJ.27:1153377:1155197:+   
6000  LinJ.29       LINF_290033800  LinJ.29:1181710:1184611:+   
6001  LinJ.29       LINF_290033800  LinJ.29:1181710:1184611:+   
6820  LinJ.31  LINF_310033100-3000  LinJ.31:1273465:1275764:-   
6821  LinJ.31  LINF_310033100-3000  LinJ.31:1273465:1275764:-   
6844  LinJ.31  LINF_310035100-5000  LinJ.31:1343999:1346787:-   
6845  LinJ.31  LINF_310035100-5000  LinJ.31:1343999:1346787:-   
7181  LinJ.32   LINF_320029000-100    LinJ.32:889966:894771:-   
7182  LinJ.32   LINF_320029000-100    LinJ.32:889966:894771:-   
7195  LinJ.32   LINF_320030400-500    LinJ.32:933016:935642:-   
7196  LinJ.32   LINF_320030400-500    LinJ.32:933016:935642:-   
7275  LinJ.32   LINF_320037300-400  LinJ.32:1177409:1179063:+   
7276  LinJ.32   LINF_320037300-400  LinJ.32:1177409:1179063:+   
7568  LinJ.33   LINF_330021700-800    LinJ.33:553571:558357:+   
7569  LinJ.33   LINF_330021700-800    LinJ.33:553571:558357:+   

             Transcript_ID           Transcript_coord    CDS/miscRNA_ID  \
777    LINF_06T0017800-900    LinJ.06:487685:490191:+    LINF_060017800   
778    LINF_06T0017800-900    LinJ.06:487685:490191:+    LINF_060017900   
841      LINF_07T0008800_1    LinJ.07:153281:154334:-  LINF_070008800_1   
842      LINF_07T0008800_2    LinJ.07:158854:160411:-  LINF_070008800_2   
1376   LINF_10T0011400-500    LinJ.10:263964:266643:+    LINF_100011400   
1377   LINF_10T0011400-500    LinJ.10:263964:266643:+    LINF_100011500   
1430     LINF_10T0016800_1    LinJ.10:468436:468904:-                ND   
1431     LINF_10T0016800_2    LinJ.10:469155:469642:-    LINF_100016800   
1748   LINF_12T0014100-200    LinJ.12:440484:443559:+    LINF_120014100   
1749   LINF_12T0014100-200    LinJ.12:440484:443559:+    LINF_120014200   
1752   LINF_12T0014500-600    LinJ.12:450370:452072:+    LINF_120014500   
1753   LINF_12T0014500-600    LinJ.12:450370:452072:+    LINF_120014600   
3774   LINF_22T0018100-200    LinJ.22:557968:563169:+    LINF_220018100   
3775   LINF_22T0018100-200    LinJ.22:557968:563169:+    LINF_220018200   
4286     LINF_24T0023900_1    LinJ.24:689866:690901:-  LINF_240023900_1   
4287     LINF_24T0023900_2    LinJ.24:692650:693427:-  LINF_240023900_2   
4469   LINF_25T0015300-400    LinJ.25:347501:349146:+    LINF_250015300   
4470   LINF_25T0015300-400    LinJ.25:347501:349146:+    LINF_250015400   
4495   LINF_25T0017700-600    LinJ.25:451512:45

We can see that some "Transcript_ID" are divided by _1 and _2, but not all. So the ones that do not end in "_1" or "_2" we should change them

In [52]:
# Let's start testing
duplicated_gene_id_df = df_filtered[df_filtered['Gene_ID'].value_counts()[df_filtered['Gene_ID']].values > 1].copy()
print(duplicated_gene_id_df.shape)
duplicated_gene_id_df

(38, 9)


Chrom              Gene_ID                 Gene_coord  \
777   LinJ.06   LINF_060017800-900    LinJ.06:487685:490191:+   
778   LinJ.06   LINF_060017800-900    LinJ.06:487685:490191:+   
841   LinJ.07       LINF_070008800    LinJ.07:153281:160411:-   
842   LinJ.07       LINF_070008800    LinJ.07:153281:160411:-   
1376  LinJ.10   LINF_100011400-500    LinJ.10:263964:266643:+   
1377  LinJ.10   LINF_100011400-500    LinJ.10:263964:266643:+   
1430  LinJ.10       LINF_100016800    LinJ.10:468436:469642:-   
1431  LinJ.10       LINF_100016800    LinJ.10:468436:469642:-   
1748  LinJ.12   LINF_120014100-200    LinJ.12:440484:443559:+   
1749  LinJ.12   LINF_120014100-200    LinJ.12:440484:443559:+   
1752  LinJ.12   LINF_120014500-600    LinJ.12:450370:452072:+   
1753  LinJ.12   LINF_120014500-600    LinJ.12:450370:452072:+   
3774  LinJ.22   LINF_220018100-200    LinJ.22:557968:563169:+   
3775  LinJ.22   LINF_220018100-200    LinJ.22:557968:563169:+   
4286  LinJ.24       LINF_240023900    LinJ.24:689866:693427:-   
4287  LinJ.24       LINF_240023900    LinJ.24:689866:693427:-   
4469  LinJ.25   LINF_250015300-400    LinJ.25:347501:349146:+   
4470  LinJ.25   LINF_250015300-400    LinJ.25:347501:349146:+   
4495  LinJ.25   LINF_250017700-600    LinJ.25:451512:453695:-   
4496  LinJ.25   LINF_250017700-600    LinJ.25:451512:453695:-   
4917  LinJ.26   LINF_260029800-900    LinJ.26:953645:955999:+   
4918  LinJ.26   LINF_260029800-900    LinJ.26:953645:955999:+   
5286  LinJ.27   LINF_270033600-700  LinJ.27:1153377:1155197:+   
5287  LinJ.27   LINF_270033600-700  LinJ.27:1153377:1155197:+   
6000  LinJ.29       LINF_290033800  LinJ.29:1181710:1184611:+   
6001  LinJ.29       LINF_290033800  LinJ.29:1181710:1184611:+   
6820  LinJ.31  LINF_310033100-3000  LinJ.31:1273465:1275764:-   
6821  LinJ.31  LINF_310033100-3000  LinJ.31:1273465:1275764:-   
6844  LinJ.31  LINF_310035100-5000  LinJ.31:1343999:1346787:-   
6845  LinJ.31  LINF_310035100-5000  LinJ.31:1343999:1346787:-   
7181  LinJ.32   LINF_320029000-100    LinJ.32:889966:894771:-   
7182  LinJ.32   LINF_320029000-100    LinJ.32:889966:894771:-   
7195  LinJ.32   LINF_320030400-500    LinJ.32:933016:935642:-   
7196  LinJ.32   LINF_320030400-500    LinJ.32:933016:935642:-   
7275  LinJ.32   LINF_320037300-400  LinJ.32:1177409:1179063:+   
7276  LinJ.32   LINF_320037300-400  LinJ.32:1177409:1179063:+   
7568  LinJ.33   LINF_330021700-800    LinJ.33:553571:558357:+   
7569  LinJ.33   LINF_330021700-800    LinJ.33:553571:558357:+   

             Transcript_ID           Transcript_coord    CDS/miscRNA_ID  \
777    LINF_06T0017800-900    LinJ.06:487685:490191:+    LINF_060017800   
778    LINF_06T0017800-900    LinJ.06:487685:490191:+    LINF_060017900   
841      LINF_07T0008800_1    LinJ.07:153281:154334:-  LINF_070008800_1   
842      LINF_07T0008800_2    LinJ.07:158854:160411:-  LINF_070008800_2   
1376   LINF_10T0011400-500    LinJ.10:263964:266643:+    LINF_100011400   
1377   LINF_10T0011400-500    LinJ.10:263964:266643:+    LINF_100011500   
1430     LINF_10T0016800_1    LinJ.10:468436:468904:-                ND   
1431     LINF_10T0016800_2    LinJ.10:469155:469642:-    LINF_100016800   
1748   LINF_12T0014100-200    LinJ.12:440484:443559:+    LINF_120014100   
1749   LINF_12T0014100-200    LinJ.12:440484:443559:+    LINF_120014200   
1752   LINF_12T0014500-600    LinJ.12:450370:452072:+    LINF_120014500   
1753   LINF_12T0014500-600    LinJ.12:450370:452072:+    LINF_120014600   
3774   LINF_22T0018100-200    LinJ.22:557968:563169:+    LINF_220018100   
3775   LINF_22T0018100-200    LinJ.22:557968:563169:+    LINF_220018200   
4286     LINF_24T0023900_1    LinJ.24:689866:690901:-  LINF_240023900_1   
4287     LINF_24T0023900_2    LinJ.24:692650:693427:-  LINF_240023900_2   
4469   LINF_25T0015300-400    LinJ.25:347501:349146:+    LINF_250015300   
4470   LINF_25T0015300-400    LinJ.25:347501:349146:+    LINF_250015400   
4495   LINF_25T0017700-600    LinJ.25:451512:45

In [53]:
# Let's loop through them by pairs
group_duplicated_gene_id_df = duplicated_gene_id_df.groupby('Gene_ID')
duplicated_gene_id_df_2 = pd.DataFrame(columns=duplicated_gene_id_df.columns)  # Create empty DataFrame to store the results

for group_name, group_df in group_duplicated_gene_id_df:
    # Check if the string in column "Transcript_ID" ends with "_1" or "_2"
    condition = group_df['Transcript_ID'].str.endswith("_1") | group_df['Transcript_ID'].str.endswith("_2") # Condition to filter
    if condition.sum() == 2:  # Since both elements have the separator "_1" and "_2", let's skip the group.
        print(f"Skipping {group_name} because it has both _1 and _2")
        duplicated_gene_id_df_2 = pd.concat([duplicated_gene_id_df_2, group_df], ignore_index=True)
        continue
    else:  # Let's differentiate the elements in "Transcript_ID" with "_1" and "_2"
        for i in range(2):
            group_df.iloc[i, :].Transcript_ID = group_df.iloc[i, :].Transcript_ID + f"_{i+1}"
        # print(group_df["Transcript_ID"])
        duplicated_gene_id_df_2 = pd.concat([duplicated_gene_id_df_2, group_df], ignore_index=True)

Skipping LINF_070008800 because it has both _1 and _2
Skipping LINF_100016800 because it has both _1 and _2
Skipping LINF_240023900 because it has both _1 and _2
Skipping LINF_290033800 because it has both _1 and _2


In [54]:
# In the original `filtered_df` replace the column "Transcript_ID" with the "Transcript_ID" values in `final_df`
# Get the index to replace in the original data frame
index_to_replace = df_filtered[df_filtered['Gene_ID'].value_counts()[df_filtered['Gene_ID']].values > 1].index

# Create a mask
mask = df_filtered.index.isin(index_to_replace)

# Replace the "Transcript_ID" in the `filtered_df` with the "Transcript_ID" in `final_df`
df_filtered.loc[mask, 'Transcript_ID'] = duplicated_gene_id_df_2.loc[:, 'Transcript_ID'].values

In [55]:
# Check results 
df_filtered.loc[mask, :]

Chrom              Gene_ID                 Gene_coord  \
777   LinJ.06   LINF_060017800-900    LinJ.06:487685:490191:+   
778   LinJ.06   LINF_060017800-900    LinJ.06:487685:490191:+   
841   LinJ.07       LINF_070008800    LinJ.07:153281:160411:-   
842   LinJ.07       LINF_070008800    LinJ.07:153281:160411:-   
1376  LinJ.10   LINF_100011400-500    LinJ.10:263964:266643:+   
1377  LinJ.10   LINF_100011400-500    LinJ.10:263964:266643:+   
1430  LinJ.10       LINF_100016800    LinJ.10:468436:469642:-   
1431  LinJ.10       LINF_100016800    LinJ.10:468436:469642:-   
1748  LinJ.12   LINF_120014100-200    LinJ.12:440484:443559:+   
1749  LinJ.12   LINF_120014100-200    LinJ.12:440484:443559:+   
1752  LinJ.12   LINF_120014500-600    LinJ.12:450370:452072:+   
1753  LinJ.12   LINF_120014500-600    LinJ.12:450370:452072:+   
3774  LinJ.22   LINF_220018100-200    LinJ.22:557968:563169:+   
3775  LinJ.22   LINF_220018100-200    LinJ.22:557968:563169:+   
4286  LinJ.24       LINF_240023900    LinJ.24:689866:693427:-   
4287  LinJ.24       LINF_240023900    LinJ.24:689866:693427:-   
4469  LinJ.25   LINF_250015300-400    LinJ.25:347501:349146:+   
4470  LinJ.25   LINF_250015300-400    LinJ.25:347501:349146:+   
4495  LinJ.25   LINF_250017700-600    LinJ.25:451512:453695:-   
4496  LinJ.25   LINF_250017700-600    LinJ.25:451512:453695:-   
4917  LinJ.26   LINF_260029800-900    LinJ.26:953645:955999:+   
4918  LinJ.26   LINF_260029800-900    LinJ.26:953645:955999:+   
5286  LinJ.27   LINF_270033600-700  LinJ.27:1153377:1155197:+   
5287  LinJ.27   LINF_270033600-700  LinJ.27:1153377:1155197:+   
6000  LinJ.29       LINF_290033800  LinJ.29:1181710:1184611:+   
6001  LinJ.29       LINF_290033800  LinJ.29:1181710:1184611:+   
6820  LinJ.31  LINF_310033100-3000  LinJ.31:1273465:1275764:-   
6821  LinJ.31  LINF_310033100-3000  LinJ.31:1273465:1275764:-   
6844  LinJ.31  LINF_310035100-5000  LinJ.31:1343999:1346787:-   
6845  LinJ.31  LINF_310035100-5000  LinJ.31:1343999:1346787:-   
7181  LinJ.32   LINF_320029000-100    LinJ.32:889966:894771:-   
7182  LinJ.32   LINF_320029000-100    LinJ.32:889966:894771:-   
7195  LinJ.32   LINF_320030400-500    LinJ.32:933016:935642:-   
7196  LinJ.32   LINF_320030400-500    LinJ.32:933016:935642:-   
7275  LinJ.32   LINF_320037300-400  LinJ.32:1177409:1179063:+   
7276  LinJ.32   LINF_320037300-400  LinJ.32:1177409:1179063:+   
7568  LinJ.33   LINF_330021700-800    LinJ.33:553571:558357:+   
7569  LinJ.33   LINF_330021700-800    LinJ.33:553571:558357:+   

               Transcript_ID           Transcript_coord    CDS/miscRNA_ID  \
777    LINF_06T0017800-900_1    LinJ.06:487685:490191:+    LINF_060017800   
778    LINF_06T0017800-900_2    LinJ.06:487685:490191:+    LINF_060017900   
841        LINF_07T0008800_1    LinJ.07:153281:154334:-  LINF_070008800_1   
842        LINF_07T0008800_2    LinJ.07:158854:160411:-  LINF_070008800_2   
1376   LINF_10T0011400-500_1    LinJ.10:263964:266643:+    LINF_100011400   
1377   LINF_10T0011400-500_2    LinJ.10:263964:266643:+    LINF_100011500   
1430       LINF_10T0016800_1    LinJ.10:468436:468904:-                ND   
1431       LINF_10T0016800_2    LinJ.10:469155:469642:-    LINF_100016800   
1748   LINF_12T0014100-200_1    LinJ.12:440484:443559:+    LINF_120014100   
1749   LINF_12T0014100-200_2    LinJ.12:440484:443559:+    LINF_120014200   
1752   LINF_12T0014500-600_1    LinJ.12:450370:452072:+    LINF_120014500   
1753   LINF_12T0014500-600_2    LinJ.12:450370:452072:+    LINF_120014600   
3774   LINF_22T0018100-200_1    LinJ.22:557968:563169:+    LINF_220018100   
3775   LINF_22T0018100-200_2    LinJ.22:557968:563169:+    LINF_220018200   
4286       LINF_24T0023900_1    LinJ.24:689866:690901:-  LINF_240023900_1   
4287       LINF_24T0023900_2    LinJ.24:692650:693427:-  LINF_240023900_2   
4469   LINF_25T0015300-400_1    LinJ.25:347501:349146:+    LINF_250015300   
4470   LINF_25T0015300-400_2    LinJ.25:347501:349146:+    LINF_250015400   
4495   LI

## Doubts with @Javi and @JMR 18/07/24 

### Doubts resolved only with @Javi

#### Change CDS coord in LINF_230015300

* New CDS coord: LinJ.23:367028:369478:-


In [56]:
# Check the element "LINF_230015300"
df_filtered[df_filtered['Gene_ID'] == "LINF_230015300"]

Chrom         Gene_ID               Gene_coord    Transcript_ID  \
3949  LinJ.23  LINF_230015300  LinJ.23:366240:369541:-  LINF_23T0015300   

             Transcript_coord  CDS/miscRNA_ID Pseudogen  \
3949  LinJ.23:366240:369541:-  LINF_230015300        NO   

            CDS/miscRNA_coord                          Function  
3949  LinJ.23:367028:369919:-  hypothetical protein - conserved

In [57]:
# Change "CDS/miscRNA_coord" to "LinJ.23:367028:369478:-"
df_filtered.loc[df_filtered['Gene_ID'] == "LINF_230015300", 'CDS/miscRNA_coord'] = "LinJ.23:367028:369478:-"

# Check the element "LINF_230015300"
df_filtered[df_filtered['Gene_ID'] == "LINF_230015300"]

Chrom         Gene_ID               Gene_coord    Transcript_ID  \
3949  LinJ.23  LINF_230015300  LinJ.23:366240:369541:-  LINF_23T0015300   

             Transcript_coord  CDS/miscRNA_ID Pseudogen  \
3949  LinJ.23:366240:369541:-  LINF_230015300        NO   

            CDS/miscRNA_coord                          Function  
3949  LinJ.23:367028:369478:-  hypothetical protein - conserved

#### Element LINF_290024900. Removing CDS

In [58]:
# Check the element "LINF_290024900"
df_filtered[df_filtered['Gene_ID'] == "LINF_290024900"]

Chrom         Gene_ID               Gene_coord    Transcript_ID  \
5906  LinJ.29  LINF_290024900  LinJ.29:850028:850959:-  LINF_29T0024900   

             Transcript_coord  CDS/miscRNA_ID Pseudogen  \
5906  LinJ.29:850028:850959:-  LINF_290024900        NO   

            CDS/miscRNA_coord                        Function  
5906  LinJ.29:849730:850788:-  paraflagellar rod protein-like

In [59]:
# Removing column "CDS/miscRNA_ID" and "CDS/miscRNA_coord" for the element "LINF_290024900" and replace it for "ND"
df_filtered.loc[df_filtered['Gene_ID'] == "LINF_290024900", ['CDS/miscRNA_ID', 'CDS/miscRNA_coord']] = "ND"

# Check the element "LINF_290024900"
df_filtered[df_filtered['Gene_ID'] == "LINF_290024900"]

Chrom         Gene_ID               Gene_coord    Transcript_ID  \
5906  LinJ.29  LINF_290024900  LinJ.29:850028:850959:-  LINF_29T0024900   

             Transcript_coord CDS/miscRNA_ID Pseudogen CDS/miscRNA_coord  \
5906  LinJ.29:850028:850959:-             ND        NO                ND   

                            Function  
5906  paraflagellar rod protein-like

#### Element LINF_350034200 and LINF_350034300

* LINF_350034200 CDS ==> Change CDS to LinJ.35:1106590:1108413:-
* LINF_350034300 ==> remove CDS

In [60]:
# Check the elements "LINF_350034200" and "LINF_350034300"
df_filtered[df_filtered['Gene_ID'].isin(["LINF_350034200", "LINF_350034300"])]

Chrom         Gene_ID                 Gene_coord    Transcript_ID  \
8710  LinJ.35  LINF_350034200  LinJ.35:1106085:1108479:-  LINF_35T0034200   
8711  LinJ.35  LINF_350034300  LinJ.35:1108768:1111622:-  LINF_35T0034300   

               Transcript_coord  CDS/miscRNA_ID Pseudogen  \
8710  LinJ.35:1106085:1108479:-  LINF_350034200        NO   
8711  LinJ.35:1108768:1111622:-  LINF_350034300        NO   

              CDS/miscRNA_coord                                     Function  
8710  LinJ.35:1109345:1111528:-  mt-LAF5|mitoribosomal LSU assembly factor 5  
8711  LinJ.35:1109345:1111843:-             Hypothetical protein - conserved

In [61]:
# Change "CDS/miscRNA_coord" to "LinJ.35:1106590:1108413:-" in "LINF_350034200"
df_filtered.loc[df_filtered['Gene_ID'] == "LINF_350034200", 'CDS/miscRNA_coord'] = "LinJ.35:1106590:1108413:-"

# Remove "CDS/miscRNA_coord" and "CDS/miscRNA_ID" in "LINF_350034300" to "ND"
df_filtered.loc[df_filtered['Gene_ID'] == "LINF_350034300", ["CDS/miscRNA_ID", "CDS/miscRNA_coord"]] = "ND"

# Check the elements "LINF_350034200" and "LINF_350034300"
df_filtered[df_filtered['Gene_ID'].isin(["LINF_350034200", "LINF_350034300"])]

Chrom         Gene_ID                 Gene_coord    Transcript_ID  \
8710  LinJ.35  LINF_350034200  LinJ.35:1106085:1108479:-  LINF_35T0034200   
8711  LinJ.35  LINF_350034300  LinJ.35:1108768:1111622:-  LINF_35T0034300   

               Transcript_coord  CDS/miscRNA_ID Pseudogen  \
8710  LinJ.35:1106085:1108479:-  LINF_350034200        NO   
8711  LinJ.35:1108768:1111622:-              ND        NO   

              CDS/miscRNA_coord                                     Function  
8710  LinJ.35:1106590:1108413:-  mt-LAF5|mitoribosomal LSU assembly factor 5  
8711                         ND             Hypothetical protein - conserved

### Doubts resolved with @Javi and @JMR

The question to @JMR about the main table was:

***
<div style="color:orange";>
Hola buenas José,

Revisando la tabla maestra hemos observado los siguientes casos:

* *Pseudogenes*: las "CDS/misRNA" coordenadas se salen de las coordenadas de "Gene_coord". Hemos pensado en eliminar el CDS.
    * **LINF_110011400**
      * Gene_coord: <u>202887</u> : 204220
      * CDS/miscRNA_coord: <u>202513</u> : 204017
  
    * **LINF_140012250**
      * Gene_coord: 248820 : <w>251321</w>
      * CDS/miscRNA_coord: 250151 : <w>251383</w>

* *Histone H4*:
  * **LINF_150005000** -→ las coordenadas "CDS/misRNA" se escapan de las coordenadas de "Gene_coord".
    * Hemos observado que el tamaño de la secuencia es de 211 nucleótidos, en comparación al resto de H4 que son de 303 nucleótidos.
    * Gene_coord: 1108 : <u>1687</u>
    * miscRNA_coord: 1505 : <u>1715</u>

* **LINF_360079700** =→ Las coordenadas del transcrito "Transcript_coord" se sobresalen de las coordenadas de "Gene_coord".
  * Gene_coord: <u>2681995</u> : <u>2683375</u>
  * Transcript_coord: <u>2681395</u> : <u>2683433</u>

¡Espero que tengáis un buen día!

R. Pacheco
</div>

***

#### Removing CDS from pseudogenes

And the answer was:
  * **LINF_110011400**: pseudogene and remove CDS
  * **LINF_140012250** pseudogenes and remove CDS

In [62]:
# Check the elements and the "CDS/miscRNA_ID" values
df_filtered[df_filtered['Gene_ID'].isin(["LINF_110011400", "LINF_140012250"])]

Chrom         Gene_ID               Gene_coord    Transcript_ID  \
1551  LinJ.11  LINF_110011400  LinJ.11:202887:204220:+  LINF_11T0011400   
2067  LinJ.14  LINF_140012250  LinJ.14:248820:251321:-  LINF_14T0012250   

             Transcript_coord  CDS/miscRNA_ID Pseudogen  \
1551  LinJ.11:202887:204220:+  LINF_110011400       YES   
2067  LinJ.14:248820:251321:-  LINF_140012250       YES   

            CDS/miscRNA_coord                             Function  
1551  LinJ.11:202513:204017:+  argonaute like protein (pseudogene)  
2067  LinJ.14:250151:251383:-                  fatty acid elongase

In [63]:
# Replacing with "ND"
df_filtered.loc[df_filtered['Gene_ID'].isin(["LINF_110011400", "LINF_140012250"]), 'CDS/miscRNA_coord'] = "ND"

# Check the elements
df_filtered[df_filtered['Gene_ID'].isin(["LINF_110011400", "LINF_140012250"])]

Chrom         Gene_ID               Gene_coord    Transcript_ID  \
1551  LinJ.11  LINF_110011400  LinJ.11:202887:204220:+  LINF_11T0011400   
2067  LinJ.14  LINF_140012250  LinJ.14:248820:251321:-  LINF_14T0012250   

             Transcript_coord  CDS/miscRNA_ID Pseudogen CDS/miscRNA_coord  \
1551  LinJ.11:202887:204220:+  LINF_110011400       YES                ND   
2067  LinJ.14:248820:251321:-  LINF_140012250       YES                ND   

                                 Function  
1551  argonaute like protein (pseudogene)  
2067                  fatty acid elongase

#### Repairing Histone H4 LINF_150005000

* New CDS coordinates: LinJ.15:1308:1505:-
* New CDS sequence: ATGGCGCGCCGCGGTGGCGTGAAGCGCATCTCGAGCGAGGTCTACGAAGAGGTGCGCCGCGTGCTGAAGGCCTACGTGGAGGACATTGTGCGCTGCAGCACGGCCTACACCGAGTACGCGCGCAAGAAGACGGTGACGGCGTGCGATGTTGTGAACGCGCTGCGCAAGCAAGGCCACATCCTCTACGGCTATGCGTAA

In [64]:
# Check the element "LINF_150005000"
df_filtered[df_filtered['Gene_ID'] == "LINF_150005000"]

Chrom         Gene_ID           Gene_coord    Transcript_ID  \
2174  LinJ.15  LINF_150005000  LinJ.15:1108:1687:-  LINF_15T0005000   

         Transcript_coord  CDS/miscRNA_ID Pseudogen    CDS/miscRNA_coord  \
2174  LinJ.15:1108:1687:-  LINF_150005000        NO  LinJ.15:1505:1715:-   

        Function  
2174  histone H4

In [65]:
# Change the coordinates from "LinJ.15:1505:1715:-" to "LinJ.15:1308:1505:-"
df_filtered.loc[df_filtered['Gene_ID'] == "LINF_150005000", 'CDS/miscRNA_coord'] = "inJ.15:1308:1505:-"

# Check the change
df_filtered[df_filtered['Gene_ID'] == "LINF_150005000"]

Chrom         Gene_ID           Gene_coord    Transcript_ID  \
2174  LinJ.15  LINF_150005000  LinJ.15:1108:1687:-  LINF_15T0005000   

         Transcript_coord  CDS/miscRNA_ID Pseudogen   CDS/miscRNA_coord  \
2174  LinJ.15:1108:1687:-  LINF_150005000        NO  inJ.15:1308:1505:-   

        Function  
2174  histone H4

#### Transcript LINF_360079700 coordinates outside gene coordinates

* New gene coordinates = LinJ.36:2681395:2683433:-
* New CDS coordinates = LinJ.36:2682378:2682647:-  (I think it was already like this).


In [66]:
# Check the element "LINF_360079700"
df_filtered[df_filtered['Gene_ID'] == "LINF_360079700"]

Chrom         Gene_ID                 Gene_coord    Transcript_ID  \
9840  LinJ.36  LINF_360079700  LinJ.36:2681995:2683375:-  LINF_36T0079700   

               Transcript_coord  CDS/miscRNA_ID Pseudogen  \
9840  LinJ.36:2681395:2683433:-  LINF_360079700        SI   

              CDS/miscRNA_coord Function  
9840  LinJ.36:2682378:2682647:-     SET8

In [67]:
# Let's change the gene coordinates from "LinJ.36:2681995:2683375:-" to "LinJ.36:2681395:2683433:-"
df_filtered.loc[df_filtered['Gene_ID'] == "LINF_360079700", 'Gene_coord'] = "LinJ.36:2681395:2683433:-"

# Check the change
df_filtered[df_filtered['Gene_ID'] == "LINF_360079700"]

Chrom         Gene_ID                 Gene_coord    Transcript_ID  \
9840  LinJ.36  LINF_360079700  LinJ.36:2681395:2683433:-  LINF_36T0079700   

               Transcript_coord  CDS/miscRNA_ID Pseudogen  \
9840  LinJ.36:2681395:2683433:-  LINF_360079700        SI   

              CDS/miscRNA_coord Function  
9840  LinJ.36:2682378:2682647:-     SET8

# Creating GTF

In [68]:
# LINF_220021000  # Removed before because it didn't have Gene_coord
javi_ignore = ["LINF_360065300"]  # #Could be bi-cistronic, need more experimental data
cds_extended_unusual_ignore = [
    # "LINF_110011400",  # Already fixed
    # "LINF_140012250",  # Already fixed
    # "LINF_150005000",   # Already fixed
    # "LINF_230015300",  # Already fixed
    # "LINF_290024900",  # Already fixed
    # "LINF_350034200",  # Already fixed
    # "LINF_350034300"  # Already fixed
    ]
transcript_extended_unusual = [
    # "LINF_360079700"  # Already fixed
    ]

# Sum every string
ignore_gene_ids = javi_ignore + cds_extended_unusual_ignore + transcript_extended_unusual

In [69]:
df_filtered[df_filtered['Gene_ID'] == "LINF_010007050"]

Chrom         Gene_ID             Gene_coord    Transcript_ID  \
21  LinJ.01  LINF_010007050  LinJ.01:54441:56024:-  LINF_01T0007050   

         Transcript_coord CDS/miscRNA_ID Pseudogen CDS/miscRNA_coord Function  
21  LinJ.01:54441:56024:-             ND        NO                ND  unknown

## Cleaning Function column

In [72]:
df_filtered["Function"].value_counts()

Function
hypothetical_protein_-_conserved                                                                                    1704
unknown                                                                                                             1101
Protein of unknown function - conserved                                                                              527
protein of unknown function - conserved                                                                              241
Hypothetical protein - conserved                                                                                     185
                                                                                                                    ... 
argininosuccinate synthase|Citrulline-aspartate ligase                                                                 1
Pteridine reductase 1|PTR1                                                                                             1
Checkpoint protein HUS1

As we can see there are a lot of names that are the same, but are divided because of case or other small characters differences

In [104]:
# Let's create a function with replace mapping to start replacing all of them.
with open("./data/tabla_maestra_function_corrector.json") as json_file:
    replace_map = json.load(json_file)

def repairing_functions_typos(replacing_map, data_frame, column_name):
    # Inverting dictionary
    inverse_replace_map = {value: key for key in replacing_map for value in replacing_map[key]}
    
    # Let's copy the data_frame
    data_frame_cp = data_frame.copy()
    
    # Let's replace "_" for " " in the column "column_name" of `data_frame_cp` 
    data_frame_cp[column_name] = data_frame_cp[column_name].str.replace("_", " ")
    
    # Applying the mapping
    data_frame_cp[column_name] = data_frame_cp[column_name].map(inverse_replace_map).fillna(data_frame_cp[column_name])
    
    return data_frame_cp

repaired_function_df = repairing_functions_typos(replace_map, df_filtered, "Function")
repaired_function_df.to_csv("./data/tabla_maestra_function_corrector.csv", index=False, header=0, sep=",")  ## To keep correcting later on
repaired_function_df['Function'].value_counts()

Function
hypothetical protein - conserved                                                 2023
unknown                                                                          1101
protein of unknown function - conserved                                           768
hypothetical protein                                                              187
protein kinase                                                                     70
                                                                                 ... 
Protein of unknown function - conserved (L1p/L10e family)                           1
DNA repair and recombination helicase protein PIF6|ATP-dependent DNA helicase       1
thymidine kinase                                                                    1
Flagellum attachment zone protein 6|FAZ6                                            1
Domain of unknown function (DUF4499)                                                1
Name: count, Length: 4438, dtype: int64

In [118]:
# Let's change df_filtered to the data in repaired_functions_df
df_filtered = repaired_function_df.copy()

## Creating GTF

In [119]:
def cds_for_gtf(gtf_row, gtf_transcript_id, gtf_transcript=False):
    gtf_feature = "CDS"
    gtf_start = gtf_row['CDS/miscRNA_coord'].split(':')[1]
    gtf_end = gtf_row['CDS/miscRNA_coord'].split(':')[2]
    if gtf_transcript:
        gtf_attribute = f'parent_id "{gtf_transcript_id}"; transcript_id "{gtf_transcript_id}"; '
    else:
        gtf_attribute = f'parent_id "{gtf_row["Gene_ID"]}"; '
    return gtf_feature, gtf_start, gtf_end, gtf_attribute

def left_utr(gtf_row, gtf_transcript_id, gtf_transcript=False):
    gtf_gene_start = int(gtf_row['Gene_coord'].split(':')[1])
    # gtf_gene_end = int(gtf_row['Gene_coord'].split(':')[2])
    gtf_sense = gtf_row['Gene_coord'].split(':')[3]

    gtf_cds_start = int(gtf_row['CDS/miscRNA_coord'].split(':')[1])
    # gtf_cds_end = int(gtf_row['CDS/miscRNA_coord'].split(':')[2])

    if gtf_gene_start != gtf_cds_start:
        gtf_feature = '3utr' if gtf_sense == '-' else '5utr'  # if the strand is "minus" the left UTR is the 3'UTR, since the transcript fo 3'utr <-----
        gtf_start = str(gtf_gene_start)
        gtf_end = str(gtf_cds_start - 1)
        gtf_strand = gtf_sense
        gtf_attribute = (
            (f'parent_id "{gtf_transcript_id}"; ' if gtf_transcript else f'parent_id "{gtf_row["Gene_ID"]}"; ')
        )
        return gtf_feature, gtf_start, gtf_end, gtf_strand, gtf_attribute
    else:
        return None, None, None, None, None


def right_utr(gtf_row, gtf_transcript_id, gtf_transcript=False):
    # utr_gene_start = int(gtf_row['Gene_coord'].split(':')[1])
    utr_gene_end = int(gtf_row['Gene_coord'].split(':')[2])
    # utr_cds_start = int(gtf_row['CDS/miscRNA_coord'].split(':')[1])
    utr_cds_end = int(gtf_row['CDS/miscRNA_coord'].split(':')[2])
    utr_sense = gtf_row['Gene_coord'].split(':')[3]

    if utr_gene_end != utr_cds_end:
        utr_feature = '5utr' if utr_sense == '-' else '3utr'
        utr_start = str(utr_cds_end + 1)
        utr_end = str(utr_gene_end)
        utr_strand = utr_sense
        utr_attribute = (
            (f'parent_id "{gtf_transcript_id}"; ' if gtf_transcript else f'parent_id "{gtf_row["Gene_ID"]}"; ')
        )
        return utr_feature, utr_start, utr_end, utr_strand, utr_attribute
    else:
        return None, None, None, None, None

In [120]:
# Let's index the dataframe and add the columns to a new list
pre_gtf = []
for index, row in df_filtered.iterrows():
    if row['Gene_ID'] in ignore_gene_ids:  # Because of the gene_id, we skip this row
        print("=" * 100)
        print(f"Skipping row {index} with Gene_ID '{row['Gene_ID']}' because it is in the ignore list")
        continue
    # GENE ID
    # noinspection SpellCheckingInspection
    seqname = row['Chrom']
    source = 'CBM'
    feature = 'gene'
    start = row['Gene_coord'].split(':')[1]
    end = row['Gene_coord'].split(':')[2]
    score = '.'
    strand = row['Gene_coord'].split(':')[3]
    frame = '.'
    attribute = (
        f'gene_id "{row['Gene_ID']}"; ' + f'gene_name "{row["Gene_ID"]}"; ' +   # gene_id needed for the gtf; gene_name for the display in IGV
        (f'pseudogen "unknown"; ' if row['Pseudogen'] in ["YES", "SI"] else "") +  # The original table have "YES" and "SI" for pseudogen :/
        (f'biotype "protein_coding"; ' if row['CDS/miscRNA_ID'] != "ND" else "") +  # biotype is protein_coding if there is a CDS/miscRNA_ID
        (f'notes "{row['Function'].replace(" ","_").replace("\t", "")}"; ' if row['Function'] != "unknown" else "") # Replace space with underscore, there can't be spaces inside the quotes
    )

    pre_gtf.append([seqname, source, feature, start, end, score, strand, frame, attribute])

    # TRANSCRIPT ID
    transcript = False
    if row['Transcript_ID'] != "ND":  # Elements that have a transcript and a CDS/miscRNA_ID
        transcript = True
        feature = 'transcript'
        start = row['Transcript_coord'].split(':')[1]
        end = row['Transcript_coord'].split(':')[2]

        transcript_id = row['Transcript_ID']
        attribute = (
            f'parent_id "{row['Gene_ID']}"; transcript_id "{transcript_id}"; transcript_name "{transcript_id}"; ' +  # transcript_id groups exon, CDS and other elements of the transcript. It is needed for the gtf display!!!!
            (f'biotype "protein_coding"; ' if row['CDS/miscRNA_ID'] != "ND" else "") + # biotype is protein_coding if there is a CDS/miscRNA_ID
            (f'notes "{row['Function'].replace(" ","_").replace("\t", "")}"; ' if row['Function'] != "unknown" else "") # Replace space with underscore, there can't be spaces inside the quotes
        )
        pre_gtf.append([seqname, source, feature, start, end, score, strand, frame, attribute])

        # CDS annotation
        if row['CDS/miscRNA_ID'] != "ND" and row['CDS/miscRNA_coord'] != "ND":  # Where the cds exists
            feature, start, end, attribute = cds_for_gtf(row, transcript_id, transcript)  # transcript is TRUE
            pre_gtf.append([seqname, source, feature, start, end, score, strand, frame, attribute]) 

            # UTR annotation
            feature, start, end, strand, attribute = left_utr(row, transcript_id, transcript)  # transcript is TRUE
            if feature is not None:
                pre_gtf.append([seqname, source, feature, start, end, score, strand, frame, attribute])

            # UTR annotation
            feature, start, end, strand, attribute = right_utr(row, transcript_id, transcript)  # transcript is TRUE
            if feature is not None:
                pre_gtf.append([seqname, source, feature, start, end, score, strand, frame, attribute])

    elif row['CDS/miscRNA_ID'] != "ND" and row['CDS/miscRNA_coord'] != "ND":  # but Transcript_ID is ND
        feature, start, end, attribute = cds_for_gtf(row, transcript_id, transcript)  # transcript is FALSE
        pre_gtf.append([seqname, source, feature, start, end, score, strand, frame, attribute]) 

        feature, start, end, strand, attribute = left_utr(row, transcript_id, transcript)  # transcript is FALSE
        if feature is not None:
            pre_gtf.append([seqname, source, feature, start, end, score, strand, frame, attribute])

        feature, start, end, strand, attribute = right_utr(row, transcript_id, transcript)  # transcript is FALSE
        if feature is not None:
            pre_gtf.append([seqname, source, feature, start, end, score, strand, frame, attribute])

Skipping row 9677 with Gene_ID 'LINF_360065300' because it is in the ignore list


In [121]:
# Create a list with the names of the columns for the GTF file:
# noinspection SpellCheckingInspection
gtf_columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']  # type: ignore

In [122]:
# Checking columns data types
gtf_df = pd.DataFrame(pre_gtf, columns=gtf_columns)
print(gtf_df.shape)
print(gtf_df.dtypes)
gtf_df.head()

(45376, 9)
seqname      object
source       object
feature      object
start        object
end          object
score        object
strand       object
frame        object
attribute    object
dtype: object


seqname source     feature start   end score strand frame  \
0  LinJ.01    CBM        gene  1520  5066     .      -     .   
1  LinJ.01    CBM  transcript  1520  5066     .      -     .   
2  LinJ.01    CBM         CDS  3710  4711     .      -     .   
3  LinJ.01    CBM        3utr  1520  3709     .      -     .   
4  LinJ.01    CBM        5utr  4712  5066     .      -     .   

                                           attribute  
0  gene_id "LINF_010005000"; gene_name "LINF_0100...  
1  parent_id "LINF_010005000"; transcript_id "LIN...  
2  parent_id "LINF_01T0005000"; transcript_id "LI...  
3                      parent_id "LINF_01T0005000";   
4                      parent_id "LINF_01T0005000";

In [123]:
# Checking two-types values for feature and strand
print(gtf_df.value_counts("feature"))
print(gtf_df.value_counts("strand"))

feature
gene          9869
transcript    9660
CDS           8744
3utr          8554
5utr          8549
Name: count, dtype: int64
strand
+    24190
-    21186
Name: count, dtype: int64


In [124]:
# Changing data types for start and end columns to integers
gtf_df[['start', 'end']] = gtf_df[['start', 'end']].apply(pd.to_numeric)
gtf_df.dtypes

seqname      object
source       object
feature      object
start         int64
end           int64
score        object
strand       object
frame        object
attribute    object
dtype: object

In [125]:
# Get duplicated rows in the data frame
gtf_df[gtf_df.duplicated(keep=False)]


seqname source feature    start      end score strand frame  \
3453   LinJ.06    CBM    gene   487685   490191     .      +     .   
3458   LinJ.06    CBM    gene   487685   490191     .      +     .   
3734   LinJ.07    CBM    gene   153281   160411     .      -     .   
3739   LinJ.07    CBM    gene   153281   160411     .      -     .   
7882   LinJ.12    CBM    gene   440484   443559     .      +     .   
7887   LinJ.12    CBM    gene   440484   443559     .      +     .   
7902   LinJ.12    CBM    gene   450370   452072     .      +     .   
7907   LinJ.12    CBM    gene   450370   452072     .      +     .   
19603  LinJ.24    CBM    gene   689866   693427     .      -     .   
19608  LinJ.24    CBM    gene   689866   693427     .      -     .   
24266  LinJ.27    CBM    gene  1153377  1155197     .      +     .   
24271  LinJ.27    CBM    gene  1153377  1155197     .      +     .   
27569  LinJ.29    CBM    gene  1181710  1184611     .      +     .   
27574  LinJ.29    CBM    gene  1181710  1184611     .      +     .   
34821  LinJ.33    CBM    gene   553571   558357     .      +     .   
34826  LinJ.33    CBM    gene   553571   558357     .      +     .   

                                               attribute  
3453   gene_id "LINF_060017800-900"; gene_name "LINF_...  
3458   gene_id "LINF_060017800-900"; gene_name "LINF_...  
3734   gene_id "LINF_070008800"; gene_name "LINF_0700...  
3739   gene_id "LINF_070008800"; gene_name "LINF_0700...  
7882   gene_id "LINF_120014100-200"; gene_name "LINF_...  
7887   gene_id "LINF_120014100-200"; gene_name "LINF_...  
7902   gene_id "LINF_120014500-600"; gene_name "LINF_...  
7907   gene_id "LINF_120014500-600"; gene_name "LINF_...  
19603  gene_id "LINF_240023900"; gene_name "LINF_2400...  
19608  gene_id "LINF_240023900"; gene_name "LINF_2400...  
24266  gene_id "LINF_270033600-700"; gene_name "LINF_...  
24271  gene_id "LINF_270033600-700"; gene_name "LINF_...  
27569  gene_id "LINF_290033800"; gene_name "LINF_2900...  
27574  gene_id "LINF_290033800"; gene_name "LINF_2900...  
34821  gene_id "LINF_330021700-800"; gene_name "LINF_...  
34826  gene_id "LINF_330021700-800"; gene_name "LINF_...

In [126]:
# This one was duplicated in the original data frame, each one with different "CDS/miscRNA_coord" values
df_filtered[df_filtered['Gene_ID'] == "LINF_270033600-700"]

Chrom             Gene_ID                 Gene_coord  \
5286  LinJ.27  LINF_270033600-700  LinJ.27:1153377:1155197:+   
5287  LinJ.27  LINF_270033600-700  LinJ.27:1153377:1155197:+   

              Transcript_ID           Transcript_coord  CDS/miscRNA_ID  \
5286  LINF_27T0033600-700_1  LinJ.27:1153377:1155197:+  LINF_270033600   
5287  LINF_27T0033600-700_2  LinJ.27:1153377:1155197:+  LINF_270033700   

     Pseudogen          CDS/miscRNA_coord                          Function  
5286        NO  LinJ.27:1153523:1154887:+  hypothetical protein - conserved  
5287        NO  LinJ.27:1153982:1154887:+  hypothetical protein - conserved

In [127]:
# Remove duplicated rows
gtf_df = gtf_df.drop_duplicates().copy()

In [128]:
# Exporting table
import csv
gtf_df.to_csv("./data/20240703111001_LINF-Tabla_maestra_v3-20244_RP_v0.8.gtf", sep="\t", index=False, header=False, quoting=csv.QUOTE_NONE, escapechar='\\')